--- a/include/sensor.hpp	2021-04-01 18:15:02.090055159 +0530
+++ b/include/sensor.hpp	2021-04-01 18:17:10.696993778 +0530
@@ -572,7 +572,8 @@
        sensorInterface->register_property("Unit", Unit);
        sensorInterface->register_property("SdrVersion", 81);
         sensorInterface->register_property("RecordType",RecordType );
-
+        sensorInterface->register_property("OwnerId",  32);
+        sensorInterface->register_property("OwnerLun", 0);
         sensorInterface->register_property("EntityId",EntityId );
         sensorInterface->register_property("EntityInstance",EntityInstance );
         sensorInterface->register_property("SensorInit",SensorInit );
