diff -Naur a/app/watchdog.cpp b/app/watchdog.cpp
--- a/app/watchdog.cpp	2020-11-30 09:54:25.883137109 -0500
+++ b/app/watchdog.cpp	2020-11-30 10:55:20.331207747 -0500
@@ -240,6 +240,9 @@
        		log<level::ERR>("Error in DontLog Set");
        		return ipmi::responseUnspecifiedError();
     	  }
+
+	uint8_t ReqRreTimeoutInterrupt = (static_cast<uint8_t>(timerPreTimeoutInterrupt) << 4) & 0xF0;
+
     try
     {
         WatchdogService wd_service;
@@ -266,6 +269,8 @@
         const uint64_t interval = initialCountdown * 100;
         wd_service.setInterval(interval);
         wd_service.resetTimeRemaining(false);
+		wd_service.setPreTimeoutInterval(preTimeoutInterval);
+		wd_service.setPreTimeoutInterrupt(ReqRreTimeoutInterrupt);
 
         // Mark as initialized so that future resets behave correctly
         wd_service.setInitialized(true);
@@ -431,7 +436,8 @@
         }
 
         // TODO: Do something about having pretimeout support
-        pretimeout = 0;
+        //pretimeout = 0;
+		pretimeout = wd_prop.preTimeoutInterval; 
 
         lastCallSuccessful = true;
         return ipmi::responseSuccess(
diff -Naur a/app/watchdog_service.cpp b/app/watchdog_service.cpp
--- a/app/watchdog_service.cpp	2020-11-30 09:54:25.883137109 -0500
+++ b/app/watchdog_service.cpp	2020-11-30 10:51:47.582365898 -0500
@@ -70,7 +70,7 @@
     }
     try
     {
-        std::map<std::string, std::variant<bool, uint64_t, std::string>>
+        std::map<std::string, std::variant<bool, uint64_t, std::string, uint8_t>>
             properties;
         response.read(properties);
         Properties wd_prop;
@@ -86,6 +86,8 @@
         wd_prop.interval = std::get<uint64_t>(properties.at("Interval"));
         wd_prop.timeRemaining =
             std::get<uint64_t>(properties.at("TimeRemaining"));
+		wd_prop.preTimeoutInterval = std::get<uint8_t>(properties.at("PreTimeoutInterval"));
+		wd_prop.preTimeoutInterrupt = std::get<uint8_t>(properties.at("PreTimeoutInterrupt"));
         return wd_prop;
     }
     catch (const std::exception& e)
@@ -198,3 +200,13 @@
 {
     setProperty("Interval", interval);
 }
+
+void WatchdogService::setPreTimeoutInterval(uint8_t pretimeoutInterval)
+{
+    setProperty("PreTimeoutInterval", pretimeoutInterval);
+}
+
+void WatchdogService::setPreTimeoutInterrupt(uint8_t preTimeoutInterrupt)
+{
+    setProperty("PreTimeoutInterrupt", preTimeoutInterrupt);
+}
diff -Naur a/app/watchdog_service.hpp b/app/watchdog_service.hpp
--- a/app/watchdog_service.hpp	2020-11-30 09:54:25.883137109 -0500
+++ b/app/watchdog_service.hpp	2020-11-30 10:52:17.700594637 -0500
@@ -38,6 +38,8 @@
         TimerUse expiredTimerUse;
         uint64_t interval;
         uint64_t timeRemaining;
+		uint8_t preTimeoutInterval;
+		uint8_t preTimeoutInterrupt;
     };
 
     /** @brief Retrieves a copy of the currently set properties on the
@@ -92,6 +94,10 @@
      */
     void setInterval(uint64_t interval);
 
+	void setPreTimeoutInterval(uint8_t pretimeoutInterval);
+
+	void setPreTimeoutInterrupt(uint8_t preTimeoutInterrupt);
+
   private:
     /** @brief sdbusplus handle */
     sdbusplus::bus::bus bus;
