diff -Naur a/dcmihandler.cpp b/dcmihandler.cpp
--- a/dcmihandler.cpp	2020-12-14 17:16:14.109544098 +0530
+++ b/dcmihandler.cpp	2020-12-14 17:16:46.494297209 +0530
@@ -1122,7 +1122,7 @@
         elog<InternalFailure>();
     }
     
-    /*
+    
     auto data = nlohmann::json::parse(sensorFile, nullptr, false);
     if (data.is_discarded())
     {
@@ -1138,7 +1138,7 @@
                         entry("POWER_SENSOR_FILE=%s", POWER_READING_SENSOR));
         elog<InternalFailure>();
     }
-    */
+    
     // Return default value if failed to read from D-Bus object
     int64_t power = 0;
     try
