From bb4305cfc11ab928ef16c32c6454bd4b7f2c5f85 Mon Sep 17 00:00:00 2001
From: eddyhan <eddyhan@ami.com.tw>
Date: Tue, 15 Dec 2020 17:18:42 +0800
Subject: [PATCH] Fix Issue 1862.


diff --git a/chassishandler.cpp b/chassishandler.cpp
index df0b59f..498a473 100644
--- a/chassishandler.cpp
+++ b/chassishandler.cpp
@@ -1143,6 +1143,22 @@ ipmi::RspType<> ipmiChassisControl(uint8_t chassisControl)
             // that it should not run since this is a direct user initiated
             // power reboot request (i.e. a reboot request that is not
             // originating via a soft power off SMS request)
+			
+			//Winnie
+			{
+				std::optional<bool> powerGood = power_policy::getPowerStatus();//Winnie
+				
+				if(!(*powerGood))
+				{	
+					log<level::ERR>("Power Good is false");
+					return ipmi::responseCommandNotAvailable();
+				}
+				else
+				{
+					log<level::ERR>("Power Good is true");
+				}
+			}
+			//Winnie
             indicate_no_softoff_needed();
 
             rc = initiate_state_transition(State::Host::Transition::Reboot);
-- 
2.7.4

