diff -Nuar a/apphandler.cpp b/apphandler.cpp
--- a/apphandler.cpp	2021-05-15 23:03:02.016879666 +0530
+++ b/apphandler.cpp	2021-05-15 23:20:32.140501623 +0530
@@ -99,7 +99,7 @@
 static constexpr const char* cmdMaskStr = "commandMask";
 static constexpr int base_16 = 16;
 #endif // ENABLE_I2C_WHITELIST_CHECK
-static constexpr uint8_t maxIPMIWriteReadSize = 144;
+static constexpr uint8_t maxIPMIWriteReadSize = 34;
 static constexpr uint8_t oemCmdStart = 192;
 static constexpr uint8_t oemCmdEnd = 255;
 static constexpr uint8_t invalidParamSelectorStart = 8;
