diff -Naur a/transporthandler.cpp b/transporthandler.cpp
--- a/transporthandler.cpp	2021-06-04 15:30:27.887026853 -0400
+++ b/transporthandler.cpp	2021-06-04 15:28:31.445868996 -0400
@@ -1872,7 +1872,7 @@
                     return responseInvalidFieldRequest();
                 }
             }
-            return response(ccParamNotSupported);
+            return responseInvalidFieldRequest();
         }
         case LanParam::MAC:
         {
