diff -Naur a/command/rakp34.cpp b/command/rakp34.cpp
--- a/command/rakp34.cpp	2020-09-09 14:17:45.869387971 +0530
+++ b/command/rakp34.cpp	2020-09-09 13:17:20.363173102 +0530
@@ -9,7 +9,8 @@
 #include <algorithm>
 #include <cstring>
 #include <phosphor-logging/log.hpp>
-
+#define NULLSTR "00000000000000000000"
+#define KEYSIZE 20
 using namespace phosphor::logging;
 
 namespace command
@@ -217,6 +218,14 @@
 
     std::copy_n(session->userName.data(), userLength, iter);
 
+    std::string chsecuritykeys = ipmi::ipmiGetChannelSecurityKeys();
+    if(strncmp(chsecuritykeys.c_str(),NULLSTR,KEYSIZE))
+    {
+    	std::fill(authAlgo->userKey.data(),
+              authAlgo->userKey.data() + authAlgo->userKey.size(), 0);
+    	std::copy_n(chsecuritykeys.c_str(), chsecuritykeys.size() , authAlgo->userKey.data());
+    }
+
     // Generate Session Integrity Key
     auto sikOutput = authAlgo->generateHMAC(input);
 
