From f537a10f41876343321796f698b85b20f3977156 Mon Sep 17 00:00:00 2001
From: vipinc <vipinc@ami.com>
Date: Thu, 22 Dec 2022 18:49:35 +0530
Subject: [PATCH] Update c++ std to 20

Signed-off-by: vipinc <vipinc@ami.com>
---
 CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index cab24e5..dd91a3d 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
 project(mail-alert-manager CXX)
-set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD 20)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/include)
-- 
2.25.1

