--- a/app/watchdog.cpp	2022-11-09 01:53:20.536882038 -0500
+++ phosphor-ipmi-host/app/watchdog.cpp	2022-11-09 01:48:47.779369377 -0500
@@ -220,7 +220,7 @@
     static constexpr auto SEN_WATCHDOG_SERVICE =
         "xyz.openbmc_project.Eventsensor";
     static constexpr auto SEN_WATCHDOG_ROOT =
-        "/xyz/openbmc_project/sensors/eventsensor/WATCHDOG2";
+        "/xyz/openbmc_project/sensors/watchdog2/IPMI_Watchdog";
     static constexpr auto WATCHDOG_SEN_INTERFACE =
         "xyz.openbmc_project.Sensor.Discrete.State";
     static constexpr auto SEN_INTERFACE = "org.freedesktop.DBus.Properties";
