diff -Naur a/src/CPUSensorMain.cpp b/src/CPUSensorMain.cpp
--- a/src/CPUSensorMain.cpp	2020-08-30 19:09:07.047721862 +0530
+++ b/src/CPUSensorMain.cpp	2020-08-30 19:13:57.731952345 +0530
@@ -304,27 +304,28 @@
             continue;
         }
 	auto findSensorUnit = baseConfiguration->second.find("Unit");
+	std::string sensorUnit;
         if (findSensorUnit == baseConfiguration->second.end())
         {
             std::cerr << "could not determine configuration Unit " << "\n";
-            continue;
         }
-        std::string sensorUnit = std::get<std::string>(findSensorUnit->second);
-
+	else
+	{
+         sensorUnit = std::get<std::string>(findSensorUnit->second);
+	}
 
         auto findEvStat = baseConfiguration->second.find("EvStat");
+	 uint16_t EvStat;
        if (findEvStat == baseConfiguration->second.end())
         {
                    std::cerr << "Base configuration missing EvStat"
             << baseConfiguration->first << "\n";
-            continue;
+            
         }
-
-        uint16_t EvStat =
-            std::visit(VariantToUnsignedIntVisitor(), findEvStat->second);
-
-
-
+       else
+	   {
+         EvStat = std::visit(VariantToUnsignedIntVisitor(), findEvStat->second);
+	   }
         // iterate through all found temp sensors
         for (const auto& inputPath : inputPaths)
         {
