diff -Naur a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2020-10-15 22:11:28.999131638 +0530
+++ b/CMakeLists.txt	2020-10-15 22:12:52.896414603 +0530
@@ -37,17 +37,17 @@
 option (YOCTO "Enable Building in Yocto" OFF)
 option (HUNTER_ENABLED "Enable hunter package pulling" OFF)
 
-option (DISABLE_ADC "Disable installing ADC sensor" OFF)
+option (DISABLE_ADC "Disable installing ADC sensor" ON)
 option (DISABLE_CPU "Disable installing CPU sensor" ON)
 option (DISABLE_DIG_GPIO "Disable intialling digital gpio sensor" ON)
 option (DISABLE_DISC_EVENT "Disable installing discrete event sensor" ON)
-option (DISABLE_EXIT_AIR "Disable installing Exit Air Temp sensor" OFF)
-option (DISABLE_FAN "Disable installing fan sensor" OFF)
+option (DISABLE_EXIT_AIR "Disable installing Exit Air Temp sensor" ON)
+option (DISABLE_FAN "Disable installing fan sensor" ON)
 option (DISABLE_HWMON_TEMP "Disable installing hwmon temp sensor" OFF)
-option (DISABLE_INTRUSION "Disable installing intrusion sensor" OFF)
-option (DISABLE_IPMB "Disable installing IPMB sensor" OFF)
-option (DISABLE_MCUTEMP "Disable installing MCU temperature sensor" OFF)
-option (DISABLE_PSU "Disable installing PSU sensor" OFF)
+option (DISABLE_INTRUSION "Disable installing intrusion sensor" ON)
+option (DISABLE_IPMB "Disable installing IPMB sensor" ON)
+option (DISABLE_MCUTEMP "Disable installing MCU temperature sensor" ON)
+option (DISABLE_PSU "Disable installing PSU sensor" ON)
 option (DISABLE_PSU_EVENT "Disable installing PSU_EVENT sensor" ON)
 option (DISABLE_NVME "Disable installing NVME sensor" ON)
 
