diff -Naur a/src/EventSensor.cpp b/src/EventSensor.cpp
--- a/src/EventSensor.cpp	2020-11-17 16:02:49.538236449 +0530
+++ b/src/EventSensor.cpp	2020-11-17 17:39:00.620987006 +0530
@@ -46,6 +46,7 @@
 
     data = {0xFF,0xFF,0xFF};
     al = 0x00;
+    uint8_t DontLog = 0;
                       discreteInterfaceState->register_property("EventData",data,
                        [&](const std::vector<uint8_t>& request, std::vector<uint8_t>& oldValue) {
                    oldValue = request;
@@ -59,6 +60,16 @@
                      al = request;
                      return 1;
                  });
+		
+		 if(!strcmp(sensorName.c_str(),"WATCHDOG2"))
+    		 {	
+		      discreteInterfaceState->register_property("DontLog",DontLog,
+                                [&](const uint8_t& request, uint8_t& oldValue) {
+                    oldValue = request;
+                     DontLog = request;
+                     return 1;
+                        });
+		 }
 
     association = objectServer.add_interface(
         "/xyz/openbmc_project/sensors/eventsensor/" + name,
