diff -Naur a/src/PwmSensor.cpp b/src/PwmSensor.cpp
--- a/src/PwmSensor.cpp	2020-12-01 12:39:33.725382709 +0530
+++ b/src/PwmSensor.cpp	2020-12-01 12:42:03.385767020 +0530
@@ -145,7 +145,15 @@
 
             return curVal;
         });
-    sensorInterface->initialize();
+
+       controlInterface->register_property(
+                "Enable", false, [this]( bool propIn, bool& old) {
+                 old = propIn;
+                sensorInterface->initialize();
+                    return 1;
+                });
+
+ //   sensorInterface->initialize();
     controlInterface->initialize();
 
     association = objectServer.add_interface(
