diff -Naur a/src/PwmSensor.cpp b/src/PwmSensor.cpp
--- a/src/PwmSensor.cpp	2021-03-01 07:41:03.960266429 +0530
+++ b/src/PwmSensor.cpp	2021-03-01 07:48:15.972093540 +0530
@@ -26,6 +26,7 @@
 
 static constexpr size_t pwmMax = 255;
 static constexpr double defaultPwm = 30.0;
+static constexpr size_t targetIfaceMax = 255;
 
 PwmSensor::PwmSensor(const std::string& name, const std::string& sysPath,
                      std::shared_ptr<sdbusplus::asio::connection>& conn,
@@ -116,7 +117,7 @@
     controlInterface->register_property(
         "Target", static_cast<uint64_t>(pwmValue),
         [this](const uint64_t& req, uint64_t& resp) {
-            if (req > pwmMax)
+            if (req > targetIfaceMax)
             {
                 throw std::runtime_error("Value out of range");
                 return -1;
@@ -125,7 +126,9 @@
             {
                 return 1;
             }
-            setValue(req);
+	     setValue(
+                std::round(pwmMax * static_cast<double>(req) / targetIfaceMax));
+
             resp = req;
 
             sensorInterface->signal_property("Value");
@@ -134,6 +137,9 @@
         },
         [this](uint64_t& curVal) {
             uint64_t value = getValue();
+	     value = static_cast<uint64_t>(std::round(
+                (static_cast<double>(value) / pwmMax) * targetIfaceMax));
+
             if (curVal != value)
             {
                 curVal = value;
