diff -Naur a/src/HwmonTempSensor.cpp b/src/HwmonTempSensor.cpp
--- a/src/HwmonTempSensor.cpp	2021-03-01 10:12:58.182808724 +0530
+++ b/src/HwmonTempSensor.cpp	2021-03-01 10:16:23.301621091 +0530
@@ -115,8 +115,18 @@
         try
         {
             double nvalue = std::stod(response);
-            nvalue /= sensorScaleFactor;
-            updateValue(nvalue);
+            if (nvalue < 0)
+            {
+                std::cerr << "Hwmon temp sensor " << name
+                          << ": ignore negative rawValue " << nvalue << "\n";
+                incrementError();
+            }
+            else
+            {
+                nvalue /= sensorScaleFactor;
+                updateValue(nvalue);
+            }
+
         }
         catch (const std::invalid_argument&)
         {
diff -Naur a/src/IpmbSensor.cpp b/src/IpmbSensor.cpp
--- a/src/IpmbSensor.cpp	2021-03-01 10:12:58.182808724 +0530
+++ b/src/IpmbSensor.cpp	2021-03-01 10:17:58.162826754 +0530
@@ -323,7 +323,10 @@
 
                 double value = 0;
 
-                if (!processReading(data, value))
+	          // Temperature sensors are not expected to read 0
+                // treat them as errors
+                if (!processReading(data, value) ||
+                    (subType == IpmbSubType::temp && value == 0.0))
                 {
                     incrementError();
                     read();
