diff -Naur a/src/NVMeSensor.cpp b/src/NVMeSensor.cpp
--- a/src/NVMeSensor.cpp	2021-03-01 10:31:45.192031649 +0530
+++ b/src/NVMeSensor.cpp	2021-03-01 10:35:14.650180325 +0530
@@ -268,7 +268,8 @@
     {
         // 0x80 = No temperature data or temperature data is more the 5 s
         // old 0x81 = Temperature sensor failure
-        return maxReading;
+        //return maxReading;
+	return std::numeric_limits<double>::quiet_NaN();
     }
 
     return reading;
@@ -368,7 +369,16 @@
                   << " Celsius for device " << sensorInfo->name << "\n";
     }
 
-    sensorInfo->updateValue(getTemperatureReading(messageData[5]));
+    double value = getTemperatureReading(messageData[5]);
+    if (!std::isfinite(value))
+    {
+        sensorInfo->markAvailable(false);
+        sensorInfo->incrementError();
+    }
+    else
+    {
+        sensorInfo->updateValue(value);
+    }
 
     if (DEBUG)
     {
