diff -Naur a/src/ADCSensor.cpp b/src/ADCSensor.cpp
--- a/src/ADCSensor.cpp	2021-03-01 19:56:07.884227785 +0530
+++ b/src/ADCSensor.cpp	2021-03-01 20:00:46.913850091 +0530
@@ -41,7 +41,7 @@
 static constexpr unsigned int sensorScaleFactor = 1000;
 
 static constexpr double roundFactor = 10000; // 3 decimal places
-static constexpr double maxReading = 20;
+static constexpr double maxReading = 1.8;    // adc reference is 1.8Volt
 static constexpr double minReading = 0;
 
 ADCSensor::ADCSensor(const std::string& path,
@@ -53,9 +53,9 @@
                      const std::string& sensorConfiguration,
                      std::optional<BridgeGpio>&& bridgeGpio,uint16_t EvStat,const std::string& sensorUnit,struct sdr::SdrRecord sdrRecord) :
     Sensor(boost::replace_all_copy(sensorName, " ", "_"),
-           std::move(_thresholds), sensorConfiguration,
-           "xyz.openbmc_project.Configuration.ADC", maxReading, minReading, EvStat, sensorUnit, sdrRecord,
-           readState),
+		    std::move(_thresholds), sensorConfiguration,
+          "xyz.openbmc_project.Configuration.ADC", maxReading / scaleFactor,
+           minReading / scaleFactor,EvStat, sensorUnit, sdrRecord, readState),
     std::enable_shared_from_this<ADCSensor>(), objServer(objectServer),
     inputDev(io, open(path.c_str(), O_RDONLY)), waitTimer(io), path(path),
     scaleFactor(scaleFactor), bridgeGpio(std::move(bridgeGpio)),
