diff -Naur a/src/ADCSensor.cpp b/src/ADCSensor.cpp
--- a/src/ADCSensor.cpp	2021-03-10 13:58:20.425173424 +0530
+++ b/src/ADCSensor.cpp	2021-03-10 14:00:07.294025387 +0530
@@ -243,6 +243,6 @@
     {
         return;
     }
-
-    thresholds::checkThresholdsPowerDelay(this, thresholdTimer);
+      thresholds::checkThresholds(this);
+//    thresholds::checkThresholdsPowerDelay(this, thresholdTimer);
 }
