--- sensors/src/Thresholds.cpp	2022-10-14 23:26:33.748421652 -0400
+++ sensors-diff/src/Thresholds.cpp	2022-10-15 01:40:17.841016340 -0400
@@ -291,6 +291,7 @@
             }
             else
             {
+		thresholdChanges.emplace_back(threshold, false, value);
                 ++cHiMidstate;
             }
         }
@@ -314,6 +315,7 @@
             }
             else
             {
+		thresholdChanges.emplace_back(threshold, false, value);
                 ++cLoMidstate;
             }
         }
