diff --git a/ipmi-whitelist.conf b/ipmi-whitelist.conf
index 5a3f154..0bd1b5c 100644
--- a/ipmi-whitelist.conf
+++ b/ipmi-whitelist.conf
@@ -350,3 +350,17 @@
 0x3e:0x75:0xff7f   //<Intel Managed Data Region>:<Get Remote Log IP>
 0x3e:0x76:0x7f7f   //<Intel Managed Data Region>:<Get/Set Password Mode>
 0x3e:0x84:0x7f7f   //<Intel Managed Data Region>:<Read PFR Mailbox>
+0x2c:0x01:0xff7f   //<Group Extension>:<Get DCMI Capability Info>
+0x2c:0x02:0xff7f   //<Group Extension>:<Get Power Reading>
+0x2c:0x03:0xff7f   //<Group Extension>:<Get Power Limit>
+0x2c:0x04:0x7f7f   //<Group Extension>:<Set Power Limit>
+0x2c:0x05:0x7f7f   //<Group Extension>:<Activate/Deactivate Power Limit>
+0x2c:0x06:0x7f7f   //<Group Extension>:<Get Asset Tag>
+0x2c:0x07:0x7f7f   //<Group Extension>:<Get DCMI Sensor Info>
+0x2c:0x08:0x7f7f   //<Group Extension>:<Set Asset Tag>
+0x2c:0x09:0x7f7f   //<Group Extension>:<Get Management Controller String Id>
+0x2c:0x0A:0x7f7f   //<Group Extension>:<Set Management Controller String Id>
+0x2c:0x10:0x7f7f   //<Group Extension>:<Get Temperature Reading>
+0x2c:0x12:0x7f7f   //<Group Extension>:<Set Dcmi Configuration Parameters>
+0x2c:0x13:0x7f7f   //<Group Extension>:<Get Dcmi Configuration Parameters>
+0x06:0x03:0x7f7f   //<App>:<Warm Reset>

