From 7c473a43d9ffa31c4230d4d704374fb9f1b62986 Mon Sep 17 00:00:00 2001
From: Ryon Heichelbech <ryonh@ami.com>
Date: Fri, 22 Oct 2021 17:13:49 -0400
Subject: [PATCH] Redfish: Add message registries for Ampere event

---
 .../registries/openbmc_message_registry.hpp   | 30 ++++++++++++++++++-
 1 file changed, 29 insertions(+), 1 deletion(-)

diff --git a/redfish-core/include/registries/openbmc_message_registry.hpp b/redfish-core/include/registries/openbmc_message_registry.hpp
index ba6f7ae..b7e41b5 100644
--- a/redfish-core/include/registries/openbmc_message_registry.hpp
+++ b/redfish-core/include/registries/openbmc_message_registry.hpp
@@ -29,7 +29,7 @@ const Header header = {
     "0.1.0",
     "OpenBMC",
 };
-constexpr std::array<MessageEntry, 284> registry = {
+constexpr std::array<MessageEntry, 286> registry = {
     MessageEntry{
         "ADDDCCorrectable",
         {
@@ -46,6 +46,34 @@ constexpr std::array<MessageEntry, 284> registry = {
             },
             "None.",
         }},
+    MessageEntry{
+        "AmpereCritical",
+        {
+            "Indicates an Ampere critical event",
+            "%1: %2",
+            "Critical",
+            "Critical",
+            2,
+            {
+                "string",
+                "string",
+            },
+            "None.",
+        }},
+    MessageEntry{
+        "AmpereWarning",
+        {
+            "Indicates an Ampere warning event",
+            "%1: %2",
+            "Warning",
+            "Warning",
+            2,
+            {
+                "string",
+                "string",
+            },
+            "None.",
+        }},
     MessageEntry{
         "AtScaleDebugFeatureEnabledAtHardware",
         {
-- 
2.17.1

