--- a/kcsbridged.cpp	2020-09-09 15:25:55.596157897 -0400
+++ b/kcsbridged.cpp	2020-09-09 15:01:39.755727729 -0400
@@ -22,6 +22,7 @@
 #include <phosphor-logging/log.hpp>
 #include <sdbusplus/asio/connection.hpp>
 #include <sdbusplus/asio/object_server.hpp>
+#include <boost/algorithm/string.hpp> 
 
 using namespace phosphor::logging;
 
@@ -110,10 +111,13 @@
         // register interfaces...
         server = std::make_shared<sdbusplus::asio::object_server>(bus);
 
+	std::vector<std::string> chnum;
+	boost::split(chnum, channel, boost::is_any_of("-"));
+
         static constexpr const char pathBase[] =
             "/xyz/openbmc_project/Ipmi/Channel/";
         std::shared_ptr<sdbusplus::asio::dbus_interface> iface =
-            server->add_interface(pathBase + channel,
+            server->add_interface(pathBase + chnum[1],
                                   "xyz.openbmc_project.Ipmi.Channel.SMS");
         iface->register_method("setAttention",
                                [this]() { return setAttention(); });
@@ -330,8 +334,10 @@
         return EXIT_FAILURE;
     }
 
+    std::vector<std::string> chnum;
+    boost::split(chnum, channel, boost::is_any_of("-")); 
     static constexpr const char busBase[] = "xyz.openbmc_project.Ipmi.Channel.";
-    std::string busName(busBase + channel);
+    std::string busName(busBase + chnum[1]);
     bus->request_name(busName.c_str());
 
     io->run();
