diff --git a/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts b/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts
index 548a809a9..6e9cb956e 100755
--- a/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts
+++ b/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts
@@ -182,15 +182,15 @@
 					};
 					kernel@200000 {
 						label = "kernel";
-						reg = <0x0200000 0x580000>;
+						reg = <0x0200000 0x600000>;
 					};
-					rofs@780000 {
+					rofs@800000 {
 						label = "rofs";
-						reg = <0x780000 0x1680000>;
+						reg = <0x800000 0x3000000>;
 					};
-					rwfs@1e00000 {
+					rwfs@38000000 {
 						label = "rwfs";
-						reg = <0x1e00000 0x200000>;
+						reg = <0x3800000 0x400000>;
 					};
 				};
 			};
