diff -Naur a/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts b/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts
--- a/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts	2020-09-24 18:20:43.543891395 +0530
+++ b/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts	2020-09-25 12:59:09.038149557 +0530
@@ -169,12 +169,11 @@
 					#size-cells = <1>;
 					bmc@0{
 						label = "bmc";
-						reg = <0x000000 0x2000000>;
+						reg = <0x000000 0x4000000>;
 					};
 					u-boot@0 {
 						label = "u-boot";
 						reg = <0x0000000 0xC0000>;
-						read-only;
 					};
 					u-boot-env@100000{
 						label = "u-boot-env";
