diff --git a/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts b/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts
old mode 100755
new mode 100644
index 84f7f8fed..f5f2d8486
--- a/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts
+++ b/arch/arm/boot/dts/nuvoton-npcm750-runbmc-olympus.dts
@@ -771,6 +771,8 @@
 					gpio-controller;
 					#gpio-cells = <2>;
 					reset-gpios = <&gpio5 28 GPIO_ACTIVE_LOW>;
+					gpio-line-names="","","","POWER_OUT","RESET_OUT","","","",
+									"","","","","","","","";
 					G2A_P0_0 {
 						gpio-hog;
 						gpios = <0 0>;
@@ -1170,3 +1172,22 @@
 	idle-states = <6>;
 	};
 };
+
+&peci0 {
+	cmd-timeout-ms = <1000>;
+	pull-down = <0>;
+	host-neg-bit-rate = <15>;
+	status = "okay";
+
+	intel-peci-dimmtemp@30 {
+		compatible = "intel,peci-client";
+		reg = <0x30>;
+		status = "okay";
+	};
+
+	intel-peci-dimmtemp@31 {
+		compatible = "intel,peci-client";
+		reg = <0x31>;
+		status = "okay";
+	};
+};
