diff -Naur a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2021-03-27 12:14:49.591375434 +0530
+++ b/CMakeLists.txt	2021-03-27 12:16:58.118795942 +0530
@@ -37,14 +37,14 @@
 option (YOCTO "Enable Building in Yocto" OFF)
 option (HUNTER_ENABLED "Enable hunter package pulling" OFF)
 
-option (DISABLE_ADC "Disable installing ADC sensor" ON)
+option (DISABLE_ADC "Disable installing ADC sensor" OFF)
 option (DISABLE_CPU "Disable installing CPU sensor" ON)
-option (DISABLE_DIG_GPIO "Disable intialling digital gpio sensor" ON)
-option (DISABLE_DISC_EVENT "Disable installing discrete event sensor" ON)
+option (DISABLE_DIG_GPIO "Disable intialling digital gpio sensor" OFF)
+option (DISABLE_DISC_EVENT "Disable installing discrete event sensor" OFF)
 option (DISABLE_EXIT_AIR "Disable installing Exit Air Temp sensor" ON)
-option (DISABLE_FAN "Disable installing fan sensor" ON)
+option (DISABLE_FAN "Disable installing fan sensor" OFF)
 option (DISABLE_HWMON_TEMP "Disable installing hwmon temp sensor" OFF)
-option (DISABLE_INTRUSION "Disable installing intrusion sensor" ON)
+option (DISABLE_INTRUSION "Disable installing intrusion sensor" OFF)
 option (DISABLE_IPMB "Disable installing IPMB sensor" ON)
 option (DISABLE_MCUTEMP "Disable installing MCU temperature sensor" ON)
 option (DISABLE_PSU "Disable installing PSU sensor" ON)
diff -Naur a/src/FanMain.cpp b/src/FanMain.cpp
--- a/src/FanMain.cpp	2021-03-27 23:44:39.590947914 +0530
+++ b/src/FanMain.cpp	2021-03-27 23:45:49.996659639 +0530
@@ -140,7 +140,7 @@
                     address = std::stoi(link.substr(findDash + 1), nullptr, 16);
                 }
                 // convert to 0 based
-                size_t index = std::stoul(indexStr) - 1;
+                size_t index = std::stoul(indexStr);
 
                 const char* baseType;
                 const SensorData* sensorData = nullptr;
@@ -397,7 +397,7 @@
                 for (const auto& [index, configPath, name] : pwmNumbers)
                 {
                     if (boost::ends_with(pwm.string(),
-                                         std::to_string(index + 1)))
+                                         std::to_string(index)))
                     {
 		 	const char* pwmbaseType;
                  	const SensorData* pwmsensorData = nullptr;
