diff -Naur e/drivers/net/phy/broadcom.c f/drivers/net/phy/broadcom.c
--- e/drivers/net/phy/broadcom.c	2020-07-30 16:06:38.244415287 -0400
+++ f/drivers/net/phy/broadcom.c	2020-07-30 16:05:24.567430921 -0400
@@ -51,6 +51,35 @@
 	phy_write(phydev, MDIO_DEVAD_NONE, MIIM_BCM54XX_EXP_DATA, value);
 }
 
+/* Broadcom BCM54612E */
+static int bcm54612e_config(struct phy_device *phydev)
+{
+	u32 reg = 0;
+
+	/* reset the PHY */
+	reg = phy_read(phydev, MDIO_DEVAD_NONE, MII_BMCR);
+	reg |= BMCR_RESET;
+	phy_write(phydev, MDIO_DEVAD_NONE, MII_BMCR, reg);
+	while (phy_read(phydev, MDIO_DEVAD_NONE, MII_BMCR) & BMCR_RESET);
+	/* phy_reset(phydev); */
+
+	/* 125Mhz Clock Output Enable */
+	reg = phy_read(phydev, MDIO_DEVAD_NONE, MIIM_BCM54XX_EXP_SEL);
+	reg |= 0xD34;
+	phy_write(phydev, MDIO_DEVAD_NONE, MIIM_BCM54XX_EXP_SEL, reg);
+
+	reg = phy_read(phydev, MDIO_DEVAD_NONE, MIIM_BCM54XX_EXP_DATA);
+	reg |= (1 << 1);
+	phy_write(phydev, MDIO_DEVAD_NONE, MIIM_BCM54XX_EXP_DATA, reg);
+
+	reg = phy_read(phydev, MDIO_DEVAD_NONE, MIIM_BCM54XX_EXP_SEL);
+	reg &= 0xfffff000;
+	phy_write(phydev, MDIO_DEVAD_NONE, MIIM_BCM54XX_EXP_SEL, reg);
+
+	genphy_config_aneg(phydev);
+	return 0;
+}
+
 /* Broadcom BCM5461S */
 static int bcm5461_config(struct phy_device *phydev)
 {
@@ -338,6 +367,27 @@
 	return bcm54xx_parse_status(phydev);
 }
 
+static int bcm5221_config(struct phy_device *phydev)
+{
+	genphy_config_aneg(phydev);
+
+	phy_reset(phydev);
+
+	return 0;
+}
+
+static int bcm5221_startup(struct phy_device *phydev)
+{
+	int ret;
+
+	/* Read the Status (2x to make sure link is right) */
+	ret = genphy_update_link(phydev);
+	if (ret)
+		return ret;
+
+	return genphy_parse_link(phydev);
+}
+
 static struct phy_driver BCM54616S_driver = {
 	.name = "Broadcom BCM54616S",
 	.uid = 0x03625d12,
@@ -388,6 +438,26 @@
 	.shutdown = &genphy_shutdown,
 };
 
+static struct phy_driver BCM5221_driver = {
+	.name = "Broadcom BCM5221",
+	.uid = 0x4061e4,
+	.mask = 0xfffff0,
+	.features = PHY_BASIC_FEATURES,
+	.config = &bcm5221_config,
+	.startup = &bcm5221_startup,
+	.shutdown = &genphy_shutdown,
+};
+
+static struct phy_driver BCM54612E_driver = {
+	.name = "Broadcom BCM54612E",
+	.uid = 0x03625e6a,
+	.mask = 0xfffff0,
+	.features = PHY_GBIT_FEATURES,
+	.config = &bcm54612e_config,
+	.startup = &bcm54xx_startup,
+	.shutdown = &genphy_shutdown,
+};
+
 int phy_broadcom_init(void)
 {
 	phy_register(&BCM54616S_driver);
@@ -395,6 +465,8 @@
 	phy_register(&BCM5464S_driver);
 	phy_register(&BCM5461S_driver);
 	phy_register(&BCM_CYGNUS_driver);
+	phy_register(&BCM5221_driver);
+	phy_register(&BCM54612E_driver);
 
 	return 0;
 }
diff -Naur e/drivers/net/phy/phy.c f/drivers/net/phy/phy.c
--- e/drivers/net/phy/phy.c	2020-07-30 16:05:08.259212518 -0400
+++ f/drivers/net/phy/phy.c	2020-07-30 16:02:04.640738190 -0400
@@ -21,6 +21,8 @@
 
 DECLARE_GLOBAL_DATA_PTR;
 
+#define debug(fmt,args...)      printf (fmt ,##args)
+
 /* Generic PHY support and helper functions */
 
 /**
