diff -Naur g/board/nuvoton/poleg/Kconfig h/board/nuvoton/poleg/Kconfig
--- g/board/nuvoton/poleg/Kconfig	2020-08-18 14:50:49.656401498 -0400
+++ h/board/nuvoton/poleg/Kconfig	2020-08-18 14:52:36.185367376 -0400
@@ -16,5 +16,15 @@
 	  The address of the Linux NIST version, which should be identical
 	  to the u-boot version.
 
+choice
+	prompt "Target board select"
+	default TARGET_POLEG_EVB
+config TARGET_POLEG_EVB
+	bool "Poleg EVB"
+
+config TARGET_POLEG_RUNBMC
+	bool "Poleg RunBMC"
+
+endchoice
 
 endif
diff -Naur g/board/nuvoton/poleg/poleg.c h/board/nuvoton/poleg/poleg.c
--- g/board/nuvoton/poleg/poleg.c	2020-08-18 14:50:49.656401498 -0400
+++ h/board/nuvoton/poleg/poleg.c	2020-08-18 14:52:39.761399799 -0400
@@ -22,6 +22,54 @@
 
 DECLARE_GLOBAL_DATA_PTR;
 
+static int board_sd_clk_init(const char *name)
+{
+	struct udevice *clk_dev;
+	struct clk clk;
+	int node, err;
+	const char *path;
+	uint clkd[2]; /* clk_id and clk_no */
+	int clk_offset;
+	int rate;
+
+	node = fdt_path_offset(gd->fdt_blob, "/aliases");
+	if (node < 0)
+		return -FDT_ERR_NOTFOUND;
+
+	path = fdt_getprop(gd->fdt_blob, node, name, NULL);
+	if (!path) {
+		printf("no alias for mmc0\n");
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	node = fdt_path_offset(gd->fdt_blob, path);
+	err = fdtdec_get_int_array(gd->fdt_blob, node, "clocks", clkd,
+			2);
+	if (err)
+		return -FDT_ERR_NOTFOUND;
+
+	rate = fdtdec_get_int(gd->fdt_blob, node, "clock-frequency", 400000);
+
+	clk_offset = fdt_node_offset_by_phandle(gd->fdt_blob, clkd[0]);
+	if (clk_offset < 0)
+		return clk_offset;
+
+	err = uclass_get_device_by_of_offset(UCLASS_CLK, clk_offset, &clk_dev);
+	if (err)
+		return err;
+
+	clk.id = clkd[1];
+	err = clk_request(clk_dev, &clk);
+	if (err < 0)
+		return err;
+
+	err = clk_set_rate(&clk, rate);
+	clk_free(&clk);
+	if (err < 0)
+		return err;
+
+	return 0;
+}
 
 static bool is_security_enabled(void)
 {
@@ -199,7 +247,7 @@
 {
 	struct npcm750_gcr *gcr = (struct npcm750_gcr *)npcm750_get_base_gcr();
 
-	int RAMsize = (readl(&gcr->intcr3) >> 8) & 0x3;
+	int RAMsize = (readl(&gcr->intcr3) >> 8) & 0x7;
 
 	switch(RAMsize)
 	{
@@ -213,9 +261,11 @@
 				gd->ram_size = 0x20000000; /* 512 MB. */
 				break;
 		case 3:
-		case 4:
 				gd->ram_size = 0x40000000; /* 1024 MB. */
 				break;
+		case 4:
+				gd->ram_size = 0x80000000; /* 2048 MB. */
+				break;
 
 		default:
 			break;
