diff -Naur a/app/access-control/controllers/user-controller.js b/app/access-control/controllers/user-controller.js
--- a/app/access-control/controllers/user-controller.js	2020-02-13 20:07:24.381948725 -0500
+++ b/app/access-control/controllers/user-controller.js	2020-02-13 20:03:38.745990295 -0500
@@ -149,7 +149,8 @@
             })
             .catch((error) => {
               console.log(JSON.stringify(error));
-              toastService.error(`Unable to update user '${originalUsername}'.`)
+              var errorStatus = JSON.parse(JSON.stringify(error));
+              toastService.error(`Unable to update user '${originalUsername}' with status ` + errorStatus.statusText)
             })
             .finally(() => {
               $scope.loading = false;
@@ -176,11 +177,12 @@
             })
             .catch((error) => {
               console.log(JSON.stringify(error));
+              var errorStatus = JSON.parse(JSON.stringify(error));
               let message;
               if (users.length > 1) {
                 message = 'Failed to remove users.'
               } else {
-                message = `Failed to remove user '${users[0].UserName}'.`
+                message = `Failed to remove user '${users[0].UserName}' with status ` + errorStatus.statusText;
               }
               toastService.error(message);
             })
