diff -Naur a/app/server-control/controllers/virtual-media-controller.js b/app/server-control/controllers/virtual-media-controller.js
--- a/app/server-control/controllers/virtual-media-controller.js	2020-11-05 15:25:11.206391584 +0530
+++ b/app/server-control/controllers/virtual-media-controller.js	2020-11-05 15:43:39.147193951 +0530
@@ -41,8 +41,22 @@
             if(instance == false) {
 	       sessionStorage.setItem("VM_Index",index);
                $scope.devices[index].isActive = true;
-               var file = $scope.devices[index].file;
-               var id = $scope.devices[index].id;
+
+	       var file = $scope.devices[index].file;
+	       var validate_file_name = document.getElementById('file-upload').value;
+
+       	       //console.log(' validate_file_name : ' + validate_file_name);
+	       var reg = /(.*?)\.(nrg|iso|img|ima|ISO|NRG|IMG|IMA)$/;
+	       if(!validate_file_name.match(reg))
+       	       {
+                   toastService.error('File_Name "' + validate_file_name +
+                                  '" It is not supported format, Please browse valid type of file..');
+    	       	$scope.resetFile(index);
+	       	$scope.stopVM(index);
+       	       }	
+
+	       else {
+	       var id = $scope.devices[index].id;
                var host = dataService.getHost().replace('https://', '');
                var token = $cookies.get('XSRF-TOKEN');
                var server =
@@ -50,6 +64,8 @@
                $scope.devices[index].nbdServer = server;
                nbdServerService.addConnection(id, server, file);
                server.start();
+		}
+		
             }
             else {
                window.alert('WARNING! Already Master Virtual Media Session is running.\n'
