--- a/app/configuration/controllers/network-controller.js	2020-12-15 11:19:20.632197402 +0800
+++ b/app/configuration/controllers/network-controller.js	2020-12-15 11:21:16.584719422 +0800
@@ -178,7 +178,14 @@
           promises.push(setMACAddress());
         }
         if($scope.defaultGateway != dataService.defaultgateway) {
-          promises.push(setDefaultGateway());
+          if( $scope.interface.DHCPEnabled )
+			{
+				  toastService.error('Unable to save network settings. Because changing the default gateway address when interface is in DHCP mode.');
+                  $scope.loading = false;
+				  return;
+			}	
+		 else
+				promises.push(setDefaultGateway());
         }
 
         // if ($scope.interface.DHCPEnabled == $scope.oldInterface.DHCPEnabled) {
