diff -Naur a/app/access-control/controllers/user-accounts-modal-settings.html b/app/access-control/controllers/user-accounts-modal-settings.html
--- a/app/access-control/controllers/user-accounts-modal-settings.html	2020-12-15 15:18:37.668396356 -0500
+++ b/app/access-control/controllers/user-accounts-modal-settings.html	2020-12-15 15:21:09.225078946 -0500
@@ -49,6 +49,7 @@
                      type="number"
                      aria-describedby="lockoutMethod1Helper"
                      ng-min="modalCtrl.settings.lockoutMethod ? 1 : null"
+                     ng-maxlength="modalCtrl.settings.lockoutMethod ? 5 : null"
                      ng-disabled="!modalCtrl.settings.lockoutMethod"
                      ng-required="modalCtrl.settings.lockoutMethod"
                      ng-model="modalCtrl.settings.timeoutDuration"/>
diff -Naur a/app/server-health/controllers/log-controller.js b/app/server-health/controllers/log-controller.js
--- a/app/server-health/controllers/log-controller.js	2020-12-15 15:18:38.036388407 -0500
+++ b/app/server-health/controllers/log-controller.js	2020-12-15 15:21:34.924516164 -0500
@@ -93,18 +93,32 @@
           };
 
           $scope.filterByDate = function(log) {
-            var endDate;
-            if ($scope.end_date &&
-                typeof $scope.end_date.getTime === 'function') {
-              endDate = new Date($scope.end_date.getTime());
-              endDate.setTime(endDate.getTime() + 86399000);
+            var logDate = new Date(log.Created);
+            var sdate = new Date($scope.start_date);
+            var edate = new Date($scope.end_date);
+
+            // Set Nan If selected date Cleared from WEBUI
+              sdate.getTime() == 0 ? NaN : sdate;
+              edate.getTime() == 0 ? NaN : edate;
+
+            if($scope.end_date && typeof $scope.end_date.getTime === 'function'){
+              edate = new Date($scope.end_date.getTime());
+              edate.setTime(edate.getTime() + 86399000);
             }
 
-            if ($scope.start_date && endDate) {
-              return (
-                  log.Timestamp >= $scope.start_date &&
-                  log.Timestamp <= endDate);
+            // If both dates are selected
+            if ($scope.start_date && $scope.end_date) {
+              return (logDate >= sdate && logDate <= edate);
             } else {
+              if($scope.start_date && !$scope.end_date){
+                // Returns the Log data from the Selected Start Date
+                return (logDate >= sdate);
+              }
+              if(!$scope.start_date && $scope.end_date){
+                // Returns the Log data upto the Selected End Date
+                return (logDate <= edate);
+              }
+              //shows all the data as default
               return true;
             }
           };
diff -Naur a/app/server-health/controllers/syslog-controller.js b/app/server-health/controllers/syslog-controller.js
--- a/app/server-health/controllers/syslog-controller.js	2020-12-15 15:18:38.036388407 -0500
+++ b/app/server-health/controllers/syslog-controller.js	2020-12-15 15:21:59.351981609 -0500
@@ -154,10 +154,31 @@
 
           $scope.filterByDate = function(log) {
             var logDate = new Date(log.Created);
+            var sdate = new Date($scope.start_date);
+            var edate = new Date($scope.end_date);
+
+            // Set Nan If selected date Cleared from WEBUI
+              sdate.getTime() == 0 ? NaN : sdate;
+              edate.getTime() == 0 ? NaN : edate;
+
+            if($scope.end_date && typeof $scope.end_date.getTime === 'function'){
+              edate = new Date($scope.end_date.getTime());
+              edate.setTime(edate.getTime() + 86399000);
+            }
+
+            // If both dates are selected
             if ($scope.start_date && $scope.end_date) {
-              return (
-                  logDate >= $scope.start_date && logDate <= $scope.end_date);
+              return (logDate >= sdate && logDate <= edate);
             } else {
+              if($scope.start_date && !$scope.end_date){
+                // Returns the Log data from the Selected Start Date
+                return (logDate >= sdate);
+              }
+              if(!$scope.start_date && $scope.end_date){
+                // Returns the Log data upto the Selected End Date
+                return (logDate <= edate);
+              }
+              //shows all the data as default
               return true;
             }
           };
