diff -Naur a/app/access-control/controllers/user-controller.js b/app/access-control/controllers/user-controller.js
--- a/app/access-control/controllers/user-controller.js	2021-05-31 21:17:04.264110268 +0530
+++ b/app/access-control/controllers/user-controller.js	2021-05-31 21:20:52.491740636 +0530
@@ -267,15 +267,11 @@
        */
       function updateAccountSettings(lockoutDuration, lockoutThreshold, password_policy_history, password_policy_complexity) {
         $scope.loading = true;
-        APIUtils.save_password_policy_history_api(password_policy_history).then(function(){
-          $scope.accountSettings['PasswordPolicyHistory'] = password_policy_history;
-        },function(){
-          toastService.error("Error in saving Password Policy History");
-        });
 
-        APIUtils.password_policy_complexity_api(password_policy_complexity).then(function(){
-          $scope.accountSettings['PasswordPolicyComplexity'] = password_policy_complexity;
-        },function(){
+        APIUtils.save_password_policy_settings(password_policy_history, password_policy_complexity).then(function(){
+          $scope.accountSettings.Oem.OpenBMC.PasswordPolicyHistory = password_policy_history;
+          $scope.accountSettings.Oem.OpenBMC.PasswordPolicyComplexity = password_policy_complexity;
+        }, function(){
           toastService.error("Error in saving Password Policy Complexity");
         });
 
@@ -315,8 +311,8 @@
                 this.settings = {};
                 this.settings.maxLogin =
                     $scope.accountSettings.AccountLockoutThreshold;
-                this.settings.PasswordPolicyHistory = $scope.accountSettings.PasswordPolicyHistory;
-                this.settings.PasswordPolicyComplexity = $scope.accountSettings.PasswordPolicyComplexity;
+                this.settings.PasswordPolicyHistory = $scope.accountSettings.Oem.OpenBMC.PasswordPolicyHistory;
+                this.settings.PasswordPolicyComplexity = $scope.accountSettings.Oem.OpenBMC.PasswordPolicyComplexity;
                 this.settings.lockoutMethod = lockoutMethod;
                 this.settings.timeoutDuration = !lockoutMethod ?
                     null :
@@ -403,7 +399,7 @@
                 this.maxPasswordLength = $scope.accountSettings ?
                     $scope.accountSettings.MaxPasswordLength :
                     null;
-                var pass_policy_complexity = $scope.accountSettings.PasswordPolicyComplexity;
+                var pass_policy_complexity = $scope.accountSettings.Oem.OpenBMC.PasswordPolicyComplexity;
                 console.log('pass_policy_complexity', pass_policy_complexity);
 
                 var that = this;
diff -Naur a/app/common/services/api-utils.js b/app/common/services/api-utils.js
--- a/app/common/services/api-utils.js	2021-05-31 21:17:04.272110255 +0530
+++ b/app/common/services/api-utils.js	2021-05-31 21:22:40.927565706 +0530
@@ -782,21 +782,17 @@
               });
         },
 
-        save_password_policy_history_api : function(password_policy_history){
+        save_password_policy_settings : function(password_policy_history, password_policy_complexity){
           console.log('save_password_policy_history_api');
+          var data = {};
+          data['Oem'] = {};
+          data.Oem['OpenBMC'] = {};
+          data.Oem.OpenBMC.PasswordPolicyComplexity = password_policy_complexity;
+          data.Oem.OpenBMC.PasswordPolicyHistory = password_policy_history
           return $http({
-            method : 'PUT',
-            url : DataService.getHost() + '/xyz/openbmc_project/user/attr/PasswordPolicyHistory',
-            data: JSON.stringify({'data': password_policy_history})
-          });
-        },
-
-        password_policy_complexity_api : function(password_policy_complexity){
-          console.log('password_policy_complexity_api');
-          return $http({
-            method : 'PUT',
-            url : DataService.getHost() + '/xyz/openbmc_project/user/attr/PasswordPolicyComplexity',
-            data: JSON.stringify({'data': password_policy_complexity})
+            method : 'PATCH',
+            url : DataService.getHost() + '/redfish/v1/AccountService',
+            data: JSON.stringify(data)
           });
         },
 
