diff -Naur a/src/bridgingcommands.cpp b/src/bridgingcommands.cpp
--- a/src/bridgingcommands.cpp	2021-04-29 20:04:46.708466953 +0530
+++ b/src/bridgingcommands.cpp	2021-04-29 20:05:12.916285491 +0530
@@ -221,7 +221,7 @@
     switch (makeCmdKey(netFn, cmd))
     {
         // Restrict ME Master write command
-        case makeCmdKey(ipmi::netFnApp, ipmi::app::cmdMasterWriteRead):
+        //case makeCmdKey(ipmi::netFnApp, ipmi::app::cmdMasterWriteRead):
         // Restrict ME OEM commands
         case makeCmdKey(netFnMeOEM, cmdMeOemSendRawPeci):
         case makeCmdKey(netFnMeOEM, cmdMeOemAggSendRawPeci):
diff -Naur a/src/manufacturingcommands.cpp b/src/manufacturingcommands.cpp
--- a/src/manufacturingcommands.cpp	2021-04-29 20:04:46.708466953 +0530
+++ b/src/manufacturingcommands.cpp	2021-04-29 20:05:12.916285491 +0530
@@ -768,7 +768,7 @@
     switch (makeCmdKey(request->ctx->netFn, request->ctx->cmd))
     {
         // i2c master write read command needs additional checking
-        case makeCmdKey(ipmi::netFnApp, ipmi::app::cmdMasterWriteRead):
+        /*case makeCmdKey(ipmi::netFnApp, ipmi::app::cmdMasterWriteRead):
             if (request->payload.size() > 4)
             {
                 // Allow write data count > 1 only in Special mode
@@ -777,7 +777,7 @@
                     return ipmi::ccInsufficientPrivilege;
                 }
             }
-            return ipmi::ccSuccess;
+            return ipmi::ccSuccess;*/
         case makeCmdKey(ipmi::netFnOemOne,
                         ipmi::intel::general::cmdGetSmSignal):
         case makeCmdKey(ipmi::netFnOemOne,
