--- git/src/storagecommands.cpp	2022-11-09 06:04:01.049747797 -0500
+++ git_new/src/storagecommands.cpp	2022-11-09 06:01:01.134126275 -0500
@@ -1278,7 +1278,7 @@
     static constexpr auto SEN_SEL_SERVICE =
         "xyz.openbmc_project.Eventsensor";
     static constexpr auto SEN_SEL_ROOT =
-        "/xyz/openbmc_project/sensors/eventsensor/SEL";
+        "/xyz/openbmc_project/sensors/event_logging_disabled/System_Event_Log";
     static constexpr auto SEN_SEL_INTERFACE =
         "xyz.openbmc_project.Sensor.Discrete.State";
     static constexpr auto SEN_SET_INTERFACE = "org.freedesktop.DBus.Properties";
