--- "a/configurations/Intel Front Panel.json"	2020-08-11 13:25:51.978095993 +0530
+++ "b/configurations/Intel Front Panel.json"	2020-08-11 13:35:05.520436751 +0530
@@ -10,6 +10,7 @@
             "Address": "0x4D",
             "Bus": "$bus",
             "Name": "Front Panel Temp",
+	    "EvStat": 231,
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
             "Thresholds": [
                 {
--- "a/configurations/1Ux16 Riser.json"	2020-08-11 13:25:51.978095993 +0530
+++ "b/configurations/1Ux16 Riser.json"	2020-08-11 13:30:01.433309547 +0530
@@ -50,6 +50,7 @@
                 "Address": "0x48",
                 "Bus": "$bus",
                 "Name": "Riser 1 Temp",
+		"EvStat": 231,
 		"Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
                 "Thresholds": [
                     {
@@ -114,6 +115,8 @@
                 "Address": "0x49",
                 "Bus": "$bus",
                 "Name": "Riser 2 Temp",
+		"EvStat": 231,
+		"Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
                 "Thresholds": [
                     {
                         "Direction": "greater than",
