--- "a/configurations/Flextronics S-1100ADU00-201 PSU.json"	2020-08-11 13:50:54.841639911 +0530
+++ "b/configurations/Flextronics S-1100ADU00-201 PSU.json"	2020-08-11 15:28:44.504223857 +0530
@@ -114,7 +114,63 @@
                         "Severity": 0,
                         "Value": 1364.8
                     },
+		    {
+                        "Direction": "less than",
+                        "Label": "fan1",
+                        "Name": "lower critical",
+                        "Severity": 1,
+                        "Value": 2080
+                    },
                     {
+                        "Direction": "less than",
+                        "Label": "fan1",
+                        "Name": "lower non critical",
+                        "Severity": 0,
+                         "Value": 2260
+                   },
+		   {
+                        "Direction": "less than",
+                        "Label": "pout1",
+                        "Name": "lower critical",
+                        "Severity": 1,
+                        "Value": 92
+                   },
+                   {
+                        "Direction": "less than",
+                        "Label": "pout1",
+                        "Name": "lower non critical",
+                        "Severity": 0,
+                        "Value": 102
+                   },
+                   {
+                        "Direction": "less than",
+                        "Label": "vin",
+                        "Name": "lower critical",
+                        "Severity": 1,
+                        "Value": 200
+                   },
+                   {
+                        "Direction": "less than",
+                        "Label": "vin",
+                        "Name": "lower non critical",
+                        "Severity": 0,
+                        "Value": 210
+                   },
+                   {
+                       "Direction": "less than",
+                       "Label": "iin",
+                       "Name": "lower critical",
+                       "Severity": 1,
+                       "Value": 1
+                   },
+                   {
+                       "Direction": "less than",
+                       "Label": "iin",
+                       "Name": "lower non critical",
+                       "Severity": 0,
+                       "Value": 2
+                   },
+                   {
                         "Direction": "greater than",
                         "Label": "iout1",
                         "Name": "upper critical",
