--- "a/configurations/WFT Baseboard.json"	2020-08-19 10:29:01.282821629 +0530
+++ "b/configurations/WFT Baseboard.json"	2020-08-19 11:13:09.261839158 +0530
@@ -248,6 +248,7 @@
             "CPURequired": 1,
             "Index": 8,
             "Name": "PVCCIN_CPU1",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Volts",
             "PowerState": "On",
             "ScaleFactor": 0.7505,
@@ -284,6 +285,7 @@
             "CPURequired": 2,
             "Index": 9,
             "Name": "PVCCIN_CPU2",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Volts",
             "PowerState": "On",
             "ScaleFactor": 0.7505,
@@ -320,6 +322,7 @@
             "CPURequired": 1,
             "Index": 14,
             "Name": "PVCCIO_CPU1",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Volts",
             "PowerState": "On",
 	    "EvStat": 231,
@@ -355,6 +358,7 @@
             "CPURequired": 2,
             "Index": 15,
             "Name": "PVCCIO_CPU2",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Volts",
 	    "EvStat": 231,
             "PowerState": "On",
@@ -390,6 +394,7 @@
             "CPURequired": 1,
             "Index": 10,
             "Name": "PVDQ_ABC_CPU1",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Volts",
             "PowerState": "On",
 	    "EvStat": 231,
@@ -425,6 +430,7 @@
             "CPURequired": 2,
             "Index": 12,
             "Name": "PVDQ_ABC_CPU2",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Volts",
             "PowerState": "On",
 	    "EvStat": 231,
@@ -460,6 +466,7 @@
             "CPURequired": 1,
             "Index": 11,
             "Name": "PVDQ_DEF_CPU1",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Volts",
             "PowerState": "On",
 	    "EvStat": 231,
@@ -495,6 +502,7 @@
             "CPURequired": 2,
             "Index": 13,
             "Name": "PVDQ_DEF_CPU2",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Volts",
             "PowerState": "On",
 	    "EvStat": 231,
@@ -703,6 +711,7 @@
             "Address": "0x8",
             "Class": "METemp",
             "Name": "SSB Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
             "PowerState": "BiosPost",
 	    "EvStat": 231,
@@ -738,6 +747,7 @@
             "Address": "0x7c",
             "Class": "PxeBridgeTemp",
             "Name": "CPU1 P12V PVCCIO VR Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -772,6 +782,7 @@
             "Address": "0x70",
             "Class": "PxeBridgeTemp",
             "Name": "CPU1 P12V PVCCIN VR Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -806,6 +817,7 @@
             "Address": "0x74",
             "Class": "PxeBridgeTemp",
             "Name": "CPU1 VR Mem ABC Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -840,6 +852,7 @@
             "Address": "0x78",
             "Class": "PxeBridgeTemp",
             "Name": "CPU1 VR Mem DEF Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -874,6 +887,7 @@
             "Address": "0x9c",
             "Class": "PxeBridgeTemp",
             "Name": "CPU2 P12V PVCCIO VR Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -908,6 +922,7 @@
             "Address": "0x50",
             "Class": "PxeBridgeTemp",
             "Name": "CPU2 P12V PVCCIN VR Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -942,6 +957,7 @@
             "Address": "0x94",
             "Class": "PxeBridgeTemp",
             "Name": "CPU2 VR Mem ABC Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -976,6 +992,7 @@
             "Address": "0x98",
             "Class": "PxeBridgeTemp",
             "Name": "CPU2 VR Mem DEF Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -1010,6 +1027,7 @@
             "Address": "0xAC",
             "Class": "PxeBridgeTemp",
             "Name": "VR P1V05 Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -1042,6 +1060,7 @@
         },
         {
             "Name": "Node Manager Sensor",
+	    "Status": "disabled",
             "Type": "NMSensor"
         },
         {
@@ -1067,6 +1086,7 @@
                 "U2_Slot4"
             ],
             "Name": "HSBP 1 Mux",
+	    "Status": "disabled",
             "Type": "PCA9546Mux"
         },
         {
@@ -1079,6 +1099,7 @@
                 "U2_Slot8"
             ],
             "Name": "HSBP 1 Mux2",
+	    "Status": "disabled",
             "Type": "PCA9546Mux"
         },
         {
@@ -1091,6 +1112,7 @@
                 "U2_Slot4"
             ],
             "Name": "HSBP 2 Mux",
+	    "Status": "disabled",
             "Type": "PCA9546Mux"
         },
         {
@@ -1103,6 +1125,7 @@
                 "U2_Slot8"
             ],
             "Name": "HSBP 2 Mux2",
+	    "Status": "disabled",
             "Type": "PCA9546Mux"
         },
         {
@@ -1115,6 +1138,7 @@
                 "U2_Slot4"
             ],
             "Name": "HSBP 3 Mux",
+	    "Status": "disabled",
             "Type": "PCA9546Mux"
         },
         {
@@ -1127,6 +1151,7 @@
                 "U2_Slot8"
             ],
             "Name": "HSBP 3 Mux2",
+	    "Status": "disabled",
             "Type": "PCA9546Mux"
         },
         {
@@ -1167,6 +1192,7 @@
                 "DIMM [ABC]\\d+ CPU1"
             ],
             "Name": "CPU1 DIMM ABC",
+	    "Status": "disabled",
             "NegativeHysteresis": 5.0,
             "OutLimitMax": 100,
             "OutLimitMin": 30,
@@ -1195,6 +1221,7 @@
                 "DIMM [DEF]\\d+ CPU1"
             ],
             "Name": "CPU1 DIMM DEF",
+	    "Status": "disabled",
             "NegativeHysteresis": 5.0,
             "OutLimitMax": 100,
             "OutLimitMin": 30,
@@ -1222,6 +1249,7 @@
                 "DIMM [ABC]\\d+ CPU2"
             ],
             "Name": "CPU2 DIMM ABC",
+	    "Status": "disabled",
             "NegativeHysteresis": 5.0,
             "OutLimitMax": 100,
             "OutLimitMin": 30,
@@ -1250,6 +1278,7 @@
                 "DIMM [DEF]\\d+ CPU2"
             ],
             "Name": "CPU2 DIMM DEF",
+	    "Status": "disabled",
             "NegativeHysteresis": 5.0,
             "OutLimitMax": 100,
             "OutLimitMin": 30,
@@ -1269,6 +1298,7 @@
         },
         {
             "Class": "temp",
+	    "Status": "disabled",
             "FFGainCoefficient": 0.0,
             "FFOffCoefficient": 0.0,
             "ICoefficient": -4.64,
@@ -1297,6 +1327,7 @@
         },
         {
             "Class": "temp",
+	    "Status": "disabled",
             "FFGainCoefficient": 0.0,
             "FFOffCoefficient": 0.0,
             "ICoefficient": -4.64,
@@ -1326,7 +1357,6 @@
         },
         {
             "Class": "temp",
-	    "Status":"disabled",
             "FFGainCoefficient": 0.0,
             "FFOffCoefficient": 0.0,
             "ICoefficient": -8.8,
@@ -1336,6 +1366,7 @@
                 "Left Rear Temp"
             ],
             "Name": "Left Rear Temp",
+	    "Status": "disabled",
             "NegativeHysteresis": 5.0,
             "OutLimitMax": 100,
             "OutLimitMin": 30,
@@ -1381,6 +1412,7 @@
         },
         {
             "Class": "temp",
+	    "Status": "disabled",
             "FFGainCoefficient": 0.0,
             "FFOffCoefficient": 0.0,
             "ICoefficient": -4.64,
@@ -1408,6 +1440,7 @@
         },
         {
             "Class": "temp",
+	    "Status": "disabled",
             "FFGainCoefficient": 0.0,
             "FFOffCoefficient": 0.0,
             "ICoefficient": -8.8,
@@ -1435,6 +1468,7 @@
         },
         {
             "Class": "temp",
+	    "Status": "disabled",
             "FFGainCoefficient": 0.0,
             "FFOffCoefficient": 0.0,
             "ICoefficient": -8.8,
@@ -1464,6 +1498,7 @@
         {
             "Address": "0x4D",
             "Bus": 6,
+	    "Status": "disabled",
             "Name": "Right Rear Board Temp",
             "Name1": "Right Rear TMP421 Internal Temp",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
@@ -1569,6 +1604,7 @@
             "Address": "0x48",
             "Bus": 6,
             "Name": "VR 1 Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
@@ -1603,6 +1639,7 @@
             "Address": "0x4B",
             "Bus": 6,
             "Name": "VR 2 Temp",
+	    "Status": "disabled",
 	    "Unit": "xyz.openbmc_project.Sensor.Value.Unit.DegreesC",
 	    "EvStat": 231,
             "Thresholds": [
