--- "a/configurations/Flextronics S-1100ADU00-201 PSU.json"	2020-08-11 16:54:34.033563337 +0530
+++ "b/configurations/Flextronics S-1100ADU00-201 PSU.json"	2020-08-11 16:55:17.297705169 +0530
@@ -5,7 +5,6 @@
                 "Address": "$address",
                 "Bus": "$bus",
                 "Name": "PSU $ADDRESS % 4 + 1 FRU",
-		"EvStat": 1,
                 "Type": "EEPROM"
             },
             {
