diff -Naur "a/configurations/Flextronics S-1100ADU00-201 PSU.json" "b/configurations/Flextronics S-1100ADU00-201 PSU.json"
--- "a/configurations/Flextronics S-1100ADU00-201 PSU.json"	2020-08-30 10:08:28.804219584 +0530
+++ "b/configurations/Flextronics S-1100ADU00-201 PSU.json"	2020-08-30 10:08:45.344622944 +0530
@@ -277,7 +277,85 @@
                     }
                 ],
                 "Type": "pmbus"
-            }
+            },
+	    {
+           "Address": "$ADDRESS % 4 + 88",
+           "Bus": "$bus",
+           "Name": "PSU$ADDRESS % 4 + 1 status",
+           "subName": "_status",
+           "EvStat": 231,
+           "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Discrete",
+           "DType": "0x08",
+           "EventType": "0x6f",
+           "EventOffset": "0x00",
+           "SubType": "0x00",
+           "Type": "pmbusdisc",
+           "SdrInfo": [
+                {
+                        "SdrVersion": "0x51",
+                        "RecordType": "0x01",
+                        "RecordLen": 0,
+                        "OwnerId": 0,
+                        "OwnerLun": "0x20",
+                        "EntityId": 0,
+                        "EntityInstance": 1,
+                        "SensorInit": "0xF7",
+                        "SensorCap": "0x48",
+                        "SensorType" : "0x02",
+                        "SensorUnit1" : 0,
+                        "SensorUnit2" : 38,
+                        "ModifierUnit": 0,
+                        "Linearization": 0,
+                        "NominalReading": "0x97",
+                        "NormalMax" : "0xC5",
+                        "NormalMin" : "0x8b",
+                        "MaxReading" : "0xff",
+                        "MinReading" : 0,
+                        "PositiveHysteris" : 0,
+                        "NegativeHysteresis" : 0,
+                        "IdStringLen" : 0
+                }
+
+        },
+   {
+           "Address": "$ADDRESS % 4 + 88",
+           "Bus": "$bus",
+           "Name": "PSU$ADDRESS % 4 + 1 fanfail",
+           "subName": "_fanfail",
+           "EvStat": 231,
+           "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Discrete",
+           "DType": "0x04",
+           "EventType": "0x03",
+           "EventOffset": "0x00",
+           "SubType": "0x01",
+           "Type": "pmbusdisc",
+           "SdrInfo": [
+                {
+
+                        "SdrVersion": "0x51",
+                        "RecordType": "0x01",
+                        "RecordLen": 0,
+                        "OwnerId": 0,
+                        "OwnerLun": "0x20",
+                        "EntityId": 0,
+                        "EntityInstance": 1,
+                        "SensorInit": "0xF7",
+                        "SensorCap": "0x48",
+                        "SensorType" : "0x02",
+                        "SensorUnit1" : 0,
+                        "SensorUnit2" : 38,
+                        "ModifierUnit": 0,
+                        "Linearization": 0,
+                        "NominalReading": "0x97",
+                        "NormalMax" : "0xC5",
+                        "NormalMin" : "0x8b",
+                       "MaxReading" : "0xff",
+                        "MinReading" : 0,
+                        "PositiveHysteris" : 0,
+                        "NegativeHysteresis" : 0,
+                        "IdStringLen" : 0
+                }
+        }
         ],
         "Name": "Flextronics S-1100ADU00-201 PSU $ADDRESS % 4 + 1",
         "Probe": "xyz.openbmc_project.FruDevice({'PRODUCT_PRODUCT_NAME': 'S-1100ADU00-201'})",
diff -Naur "a/configurations/WFT Baseboard.json" "b/configurations/WFT Baseboard.json"
--- "a/configurations/WFT Baseboard.json"	2020-08-30 10:08:28.804219584 +0530
+++ "b/configurations/WFT Baseboard.json"	2020-08-30 10:08:45.344622944 +0530
@@ -2610,21 +2610,46 @@
 
             "Type": "Discevent"
         },
+               {
+            "Unit": "xyz.openbmc_project.Sensor.Value.Unit.Discrete",
+            "EvStat": 231,
+            "Name": "SEL",
+            "DType": "0x10",
+            "EventOffset": "0",
+            "SdrInfo": [
+                {
+                        "SdrVersion": "0x51",
+                        "RecordType": "0x01",
+                        "RecordLen": 0,
+                        "OwnerId": 0,
+                        "OwnerLun": "0x20",
+                        "EntityId": "0x06",
+                        "EntityInstance": 1,
+                        "SensorInit": "0xF7",
+                        "SensorCap": "0x48",
+                        "SensorType" : "0x02",
+                       "EventType" : "0x6f",
+                        "SensorUnit1" : 0,
+                        "SensorUnit2" : 38,
+                        "ModifierUnit": 0,
+                        "Linearization": 0,
+                        "NominalReading": "0x97",
+                        "NormalMax" : "0xC5",
+                        "NormalMin" : "0x8b",
+                        "MaxReading" : "0xff",
+                        "MinReading" : 0,
+                        "PositiveHysteris" : 0,
+                        "NegativeHysteresis" : 0,
+                        "IdStringLen" : 0
+                }
+            ],
+               "Type": "Discevent"
+        },
          {
             "BindGpioIntrusion": "Chassis Intrusion GPIO",
             "Class": "Gpio",
             "Name": "Chassis Intrusion Sensor",
-            "Type": "ChassisIntrusionSensor"
-        },
-        {
-            "Direction": "Input",
-            "Index": 11,
-            "Name": "Chassis Intrusion GPIO",
-            "Polarity": "Low",
-            "DType": "0x05",
-            "EventType": "0x6f",
-            "EventOffset": "0",
-            "EvStat": 231,
+	    "Type": "ChassisIntrusionSensor",
 	    "SdrInfo": [
                 {
                         "SdrVersion": "0x51",
@@ -2637,6 +2662,7 @@
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
+			"EventType": "0x6f",
                         "SensorUnit1" : 0,
                         "SensorUnit2" : 38,
                         "ModifierUnit": 0,
@@ -2650,11 +2676,20 @@
                         "NegativeHysteresis" : 0,
                         "IdStringLen" : 0
                 }
-            ],
-
+            ]
+	 },
+        {
+            "Direction": "Input",
+            "Index": 11,
+            "Name": "Chassis Intrusion GPIO",
+            "Polarity": "Low",
+           "DType": "0x05",
+            "EventOffset": "0",
+            "EvStat": 231,
             "Type": "Gpio"
          }
 
+
     ],
     "Name": "WFP Baseboard",
     "Probe": "xyz.openbmc_project.FruDevice({'PRODUCT_PRODUCT_NAME': '.*WFT'})",
