diff -Naur a/include/configs/aspeed-common.h b/include/configs/aspeed-common.h
--- a/include/configs/aspeed-common.h	2021-03-31 17:30:11.888977354 +0530
+++ b/include/configs/aspeed-common.h	2021-03-31 17:30:35.036898291 +0530
@@ -52,6 +52,7 @@
  * Miscellaneous configurable options
  */
 #define CONFIG_SYS_BOOTM_LEN 		(0x800000 * 2)
+#define CONFIG_ENV_OVERWRITE
 
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	"verify=yes\0"	\
