--- /redfish-core/lib/account_service.hpp	2020-04-07 19:22:37.924189182 -0400
+++ /redfish-core/lib/account_service.hpp	2020-04-07 19:10:04.572441695 -0400
@@ -1457,9 +1457,8 @@
                                     getRoleIdFromPrivilege(*userPrivPtr);
                                 if (role.empty())
                                 {
-                                    BMCWEB_LOG_ERROR << "Invalid user role";
-                                    messages::internalError(asyncResp->res);
-                                    return;
+				    //ipmi allows for incomplete users with no user privilage
+                                    BMCWEB_LOG_DEBUG << "No user role";
                                 }
                                 asyncResp->res.jsonValue["RoleId"] = role;
 
