diff -Naur a/redfish-core/lib/log_services.hpp b/redfish-core/lib/log_services.hpp
--- a/redfish-core/lib/log_services.hpp	2020-08-18 18:29:19.347473480 +0530
+++ b/redfish-core/lib/log_services.hpp	2020-08-18 18:31:42.087348035 +0530
@@ -599,8 +599,11 @@
         asyncResp->res.jsonValue["@odata.type"] =
             "#LogService.v1_1_0.LogService";
         asyncResp->res.jsonValue["Name"] = "Event Log Service";
+        res.jsonValue["DateTime"] = crow::utility::dateTimeNow();
         asyncResp->res.jsonValue["Description"] = "System Event Log Service";
         asyncResp->res.jsonValue["Id"] = "EventLog";
+        asyncResp->res.jsonValue["LogEntryType"] = "Event";
+        asyncResp->res.jsonValue["MaxNumberOfRecords"] = 150;
         asyncResp->res.jsonValue["OverWritePolicy"] = "WrapsWhenFull";
         asyncResp->res.jsonValue["Entries"] = {
             {"@odata.id",
