--- a/redfish-core/lib/log_services.hpp	2020-08-26 05:54:30.902532294 +0530
+++ b/redfish-core/lib/log_services.hpp	2020-08-26 05:56:32.578595721 +0530
@@ -430,7 +430,7 @@
 static bool
     getRedfishLogFiles(std::vector<std::filesystem::path>& redfishLogFiles)
 {
-    static const std::filesystem::path redfishLogDir = "/var/log";
+    static const std::filesystem::path redfishLogDir = "/var/sellog";
     static const std::string redfishLogFilename = "redfish";
 
     // Loop through the directory looking for redfish log files
