diff -Naur a/redfish-core/lib/sensors.hpp b/redfish-core/lib/sensors.hpp
--- a/redfish-core/lib/sensors.hpp	2020-10-16 15:19:57.803940111 +0530
+++ b/redfish-core/lib/sensors.hpp	2020-10-16 15:21:25.007766007 +0530
@@ -886,8 +886,24 @@
     }
 
     sensor_json["Status"]["State"] = getState(inventoryItem);
-    sensor_json["Status"]["Health"] =
-        getHealth(sensor_json, interfacesDict, inventoryItem);
+
+    //added changes for Health NA status
+    auto stat = valueIt->second.find("SensorStat");
+    // If a stat is true Health is not applicable
+    if (stat != valueIt->second.end())
+    {
+        const bool* sensor_stat = std::get_if<bool>(&stat->second);
+        if (sensor_stat != nullptr && (*sensor_stat == true))
+        {
+            sensor_json["Status"]["Health"] = "NA";
+        }
+    }
+
+    if (sensor_json["Status"]["Health"] != "NA")
+    {
+         sensor_json["Status"]["Health"] =
+         getHealth(sensor_json, interfacesDict, inventoryItem);
+    }
 
     // Parameter to set to override the type we get from dbus, and force it to
     // int, regardless of what is available.  This is used for schemas like fan,
