diff -Naur a/redfish-core/lib/sensors.hpp b/redfish-core/lib/sensors.hpp
--- a/redfish-core/lib/sensors.hpp	2020-10-16 15:22:30.699639130 +0530
+++ b/redfish-core/lib/sensors.hpp	2020-10-16 15:25:48.659274150 +0530
@@ -1115,7 +1115,9 @@
                     }
                     else
                     {
-                       sensor_json[key] = temp;
+		       double rounded = static_cast<int> (temp * 1000.0)/1000.0;
+                       sensor_json[key] = rounded;
+                       //sensor_json[key] = temp;
                     }
 		}
             }
