From 45cad993bb86f037722d4b375192b04217fcbb02 Mon Sep 17 00:00:00 2001
From: pranoygoru <gorumanip@amiindia.co.in>
Date: Tue, 30 Mar 2021 15:01:11 +0530
Subject: [PATCH] Retrive back Serial interface patch

---
 redfish-core/include/redfish.hpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/redfish-core/include/redfish.hpp b/redfish-core/include/redfish.hpp
index efb12ae..35943e3 100644
--- a/redfish-core/include/redfish.hpp
+++ b/redfish-core/include/redfish.hpp
@@ -73,6 +73,8 @@ class RedfishService
         nodes.emplace_back(std::make_unique<SessionService>(app));
         nodes.emplace_back(std::make_unique<EthernetCollection>(app));
         nodes.emplace_back(std::make_unique<EthernetInterface>(app));
+	nodes.emplace_back(std::make_unique<SerialInterfaceCollection>(app));
+	nodes.emplace_back(std::make_unique<SerialInterfaceInstance>(app));
         nodes.emplace_back(std::make_unique<Thermal>(app));
         nodes.emplace_back(std::make_unique<ManagerCollection>(app));
         nodes.emplace_back(std::make_unique<Manager>(app));
