diff -Naur b/redfish-core/include/registries/base_message_registry.hpp a/redfish-core/include/registries/base_message_registry.hpp
--- b/redfish-core/include/registries/base_message_registry.hpp	2020-12-01 19:02:06.562506040 +0530
+++ a/redfish-core/include/registries/base_message_registry.hpp	2020-12-01 18:52:56.042665682 +0530
@@ -24,7 +24,7 @@
 {
 const Header header = {
     "Copyright 2014-2019 DMTF. All rights reserved.",
-    "#MessageRegistry.v1_3_1.MessageRegistry",
+    "#MessageRegistry.v1_4_0.MessageRegistry",
     "Base.1.7.0",
     "Base Message Registry",
     "en",
diff -Naur b/redfish-core/include/registries/openbmc_message_registry.hpp a/redfish-core/include/registries/openbmc_message_registry.hpp
--- b/redfish-core/include/registries/openbmc_message_registry.hpp	2020-12-01 19:02:06.562506040 +0530
+++ a/redfish-core/include/registries/openbmc_message_registry.hpp	2020-12-01 18:52:56.042665682 +0530
@@ -20,7 +20,7 @@
 {
 const Header header = {
     "Copyright 2018 OpenBMC. All rights reserved.",
-    "#MessageRegistry.v1_3_1.MessageRegistry",
+    "#MessageRegistry.v1_4_0.MessageRegistry",
     "OpenBMC.0.1.0",
     "OpenBMC Message Registry",
     "en",
@@ -2702,8 +2702,8 @@
                  {
                      "Indicates that a physical security event",
                      "%1 sensor of type Physical_Security logged a general chassis intrusion",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2712,8 +2712,8 @@
                  {
 	             "Indicates that a physical security event",
                      "%1 sensor of type Physical_Security logged a Drive Bay intrusion",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2722,8 +2722,8 @@
                  {
                      "Indicates that a physical security event",
 		     "%1 sensor of type Physical_Security logged a IO card area intrusion ",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2732,8 +2732,8 @@
                  {
 		     "Indicates that a physical security event",	 
                      "%1 sensor of type Physical_Security logged a processor area intrusion",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2742,8 +2742,8 @@
                  {
 		     "Indicates that a physical security event",	  
                      "%1 sensor of type Physical_Security logged a Unauthorized Dock",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2752,8 +2752,8 @@
                  {
 		     "Indicates that a physical security event",	  
                      "%1 sensor of type Physical_Security logged a Fan Area Intrusion",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2763,8 +2763,8 @@
                  {
                      "Indicate that to  disable the logging",
                      "%1 sensor of type Event Logging Disabled logged a Correctable memory error logging disabled",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2774,8 +2774,8 @@
                  {
                      "Indicate that to  disable the logging",
                      "%1 sensor of type Event Logging Disabled logged a Event logging disabled",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2785,8 +2785,8 @@
                  {
                      "Indicate that to  disable the logging",
                      "%1 sensor of type Event Logging Disabled logged a Log area reset/cleared",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2796,8 +2796,8 @@
                  {
                      "Indicate that to  disable the logging",
                      "%1 sensor of type Event Logging Disabled logged a All event logging disabled",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2807,8 +2807,8 @@
                  {
                      "Indicate that to  disable the logging",
                      "%1 sensor of type Event Logging Disabled logged a Log full",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2818,8 +2818,8 @@
                  {
                      "Indicate to disable the logging",
                      "%1 sensor of type Event Logging Disabled logged a Log Almost full",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2829,8 +2829,8 @@
                  {
                      "Indicates that there is a host watchdog event",
                      "%1 sensor of type watchdog_2 logged a Timer Expired",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2840,8 +2840,8 @@
                  {
                      "Indicates that there is a host watchdog event",
                      "%1 sensor of type watchdog_2 logged a Hard Reset",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2851,8 +2851,8 @@
                  {
                      "Indicates that there is a host watchdog event",
                      "%1 sensor of type watchdog_2 logged a Power Down",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2862,8 +2862,8 @@
                  {
                      "Indicates that there is a host watchdog event",
                      "%1 sensor of type watchdog_2 logged a Power cycle",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2873,8 +2873,8 @@
                  {
                      "Indicates that there is a host watchdog event",
                      "%1 sensor of type watchdog_2 logged a Timer interrupt ",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2884,8 +2884,8 @@
                  {
                      "Indicates that there is a host watchdog event",
                      "%1 sensor of type watchdog_2 logged a reserved ",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2918,8 +2918,8 @@
                  {
                      "Indicates that to monitor the state of peripherals",
                      "%1 sensor of type System Acpi Power State logged a S2: sleeping, processor context lost",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2929,8 +2929,8 @@
                  {
                      "Indicates that to monitor the state of peripherals",
                      "%1 sensor of type System Acpi Power State logged a S3: sleeping, processor & hw context lost, memory retained",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2951,8 +2951,8 @@
                  {
                      "Indicates that to monitor the state of peripherals",
                      "%1 sensor of type System Acpi Power State logged a S5/G2: soft-off",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2962,8 +2962,8 @@
                  {
                      "Indicates that to monitor the state of peripherals",
                      "%1 sensor of type System Acpi Power State logged a S4/S5: soft-off",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2973,8 +2973,8 @@
                  {
                      "Indicates that to monitor the state of peripherals",
                      "%1 sensor of type System Acpi Power State logged a G3: mechanical off",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2984,8 +2984,8 @@
                  {
                      "Indicates that to monitor the state of peripherals",
                      "%1 sensor of type System Acpi Power State logged a Sleeping in S1/S2/S3 state",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -2995,8 +2995,8 @@
                  {
                      "Indicates that to monitor the state of peripherals",
                      "%1 sensor of type System Acpi Power State logged a G1: sleeping",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3049,8 +3049,8 @@
                  {
                      "Indicates an correctable Memory ECC error.",
                      "%1 sensor of type Memory logged a correctable ECC",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3089,8 +3089,8 @@
                  {
                      "Indicates that a memory type sensor logged Memory Device Disabled",
                      "%1 sensor of type Memory logged a Memory Device Disabled",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3099,8 +3099,8 @@
                  {
                      "Indicates that a memory type sensor logged a  Correctable ECC logging",
                      "%1 sensor of type Memory logged a Memory logged a Correctable ECC logging limit reached",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3109,8 +3109,8 @@
                  {
                      "Indicates that a memory type sensor logged a Presence Detected",
                      "%1 sensor of type Memory logged a Memory logged a Presence Detected",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3129,8 +3129,8 @@
                  {
                      "Indicates that a memory type sensor logged a Spare ",
                      "%1 sensor of type Memory logged a Memory logged a Spare",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3139,8 +3139,8 @@
                  {
                      "Indicates that a memory type sensor logged a Throttled ",
                      "%1 sensor of type Memory logged a Memory logged a Throttled",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3149,8 +3149,8 @@
                  {
                      "Indicates that a memory type sensor logged a Critical Overtemperature ",
                      "%1 sensor of type Memory logged a Memory logged a Critical Overtemperature",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3169,8 +3169,8 @@
                  {
                      "Indicates that the drive is faulty ",
                      "%1 sensor of type Drive Slot logged a Drive Fault",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3189,8 +3189,8 @@
                  {
                      "Indicates that the senosr of type Drive slot logged HotSpare",
                      "%1 sensor of type Drive Slot logged a Hot Spare",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3199,8 +3199,8 @@
                  {
                      "Indicates that the senosr of type Drive slot Parity Check In Progress",
                      "%1 sensor of type Drive Slot logged a Parity Check In Progress",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3209,8 +3209,8 @@
                  {
                      "Indicates that the senosr of type Drive slot logged In Critical Array",
                      "%1 sensor of type Drive Slot logged  In Critical Array",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3219,8 +3219,8 @@
                  {
                      "Indicates that the senosr of type Drive slot logged In Failed Array",
                      "%1 sensor of type Drive Slot logged  In Failed Array",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3229,8 +3229,8 @@
                  {
                      "Indicates that the senosr of type Drive slot logged Rebuild In Progress",
                      "%1 sensor of type Drive Slot logged a Rebuild In Progress",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
@@ -3239,8 +3239,8 @@
                  {
                      "Indicates that the senosr of type Drive slot logged Rebuild Aborted",
                      "%1 sensor of type Drive Slot logged a Rebuild Aborted",
-                     "WARNING",
-                     "WARNING",
+                     "Warning",
+                     "Warning",
                      1,
                      {"string"},
                      "None.",
diff -Naur b/redfish-core/include/registries/task_event_message_registry.hpp a/redfish-core/include/registries/task_event_message_registry.hpp
--- b/redfish-core/include/registries/task_event_message_registry.hpp	2020-12-01 19:02:06.562506040 +0530
+++ a/redfish-core/include/registries/task_event_message_registry.hpp	2020-12-01 18:52:56.042665682 +0530
@@ -25,7 +25,7 @@
 const Header header = {
     "Copyright 2014-2018 DMTF in cooperation with the Storage Networking "
     "Industry Association (SNIA). All rights reserved.",
-    "#MessageRegistry.v1_3_1.MessageRegistry",
+    "#MessageRegistry.v1_4_0.MessageRegistry",
     "TaskEvent.1.0.1",
     "Task Event Message Registry",
     "en",
diff -Naur b/redfish-core/lib/cpudimm.hpp a/redfish-core/lib/cpudimm.hpp
--- b/redfish-core/lib/cpudimm.hpp	2020-12-01 19:02:06.562506040 +0530
+++ a/redfish-core/lib/cpudimm.hpp	2020-12-01 18:52:56.046665716 +0530
@@ -168,7 +168,7 @@
 
 
 		const std::string &processorName = params[0];
-		res.jsonValue["@odata.type"] = "#Processor.v1_3_1.Processor";
+		res.jsonValue["@odata.type"] = "#Processor.v1_7_0.Processor";
 		res.jsonValue["@odata.context"] =
 			"/redfish/v1/$metadata#Processor.Processor";
 		res.jsonValue["@odata.id"] =
@@ -349,12 +349,12 @@
 									return;
 									}
 
-									nlohmann::json& entriesArray = aResp->res.jsonValue["ConnectedProcessors"];
+									nlohmann::json& entriesArray = aResp->res.jsonValue["Links"]["ConnectedProcessors"];
 
 									std::string  core_type = "Core";
 									if((core_type.compare(*processortype)) == 0)
 									{
-										entriesArray.push_back({{"Link","/redfish/v1/Systems/system/Processors/" + processorName + "/SubProcessors/" + subprocessorName}});
+										entriesArray.push_back({{"@odata.id","/redfish/v1/Systems/system/Processors/" + processorName + "/SubProcessors/" + subprocessorName}});
 									}
 
 
@@ -819,13 +819,14 @@
 			const std::string &processorName = params[0];
 			const std::string &subprocessorName = params[1];
 
-			res.jsonValue["@odata.type"] = "#Processor.v1_3_1.Processor";
+			res.jsonValue["@odata.type"] = "#Processor.v1_7_0.Processor";
 			res.jsonValue["@odata.context"] =
 				"/redfish/v1/$metadata#Processor.Processor";
 			res.jsonValue["@odata.id"] =
 				"/redfish/v1/Systems/system/Processors/" + processorName + "/SubProcessors/" + subprocessorName;
 
 			res.jsonValue["Name"] = subprocessorName;
+			res.jsonValue["Id"] = subprocessorName;
 
 			//####################################added from here
 
diff -Naur b/redfish-core/lib/power.hpp a/redfish-core/lib/power.hpp
--- b/redfish-core/lib/power.hpp	2020-12-01 19:02:06.562506040 +0530
+++ a/redfish-core/lib/power.hpp	2020-12-01 18:52:56.046665716 +0530
@@ -234,7 +234,7 @@
                         // Mandatory properties odata.id and MemberId
                         // A warning without a odata.type
                         tempArray.push_back(
-                            {{"@odata.type", "#Power.v1_6_0.PowerControl"},
+                            {
                              {"@odata.id", "/redfish/v1/Chassis/" +
                                                sensorAsyncResp->chassisId +
                                                "/Power#/PowerControl/0"},
diff -Naur b/redfish-core/lib/sensors.hpp a/redfish-core/lib/sensors.hpp
--- b/redfish-core/lib/sensors.hpp	2020-12-01 19:02:06.562506040 +0530
+++ a/redfish-core/lib/sensors.hpp	2020-12-01 18:52:56.046665716 +0530
@@ -930,7 +930,7 @@
     else if (sensorType == "temperature")
     {
         unit = "/ReadingCelsius"_json_pointer;
-        sensor_json["@odata.type"] = "#Thermal.v1_6_0.Temperature";
+        //sensor_json["@odata.type"] = "#Thermal.v1_6_0.Temperature";
     	sensor_json["PhysicalContext"] = "Intake";
         // TODO(ed) Documentation says that path should be type fan_tach,
         // implementation seems to implement fan
@@ -939,7 +939,7 @@
     {
         unit = "/Reading"_json_pointer;
         sensor_json["ReadingUnits"] = "RPM";
-        sensor_json["@odata.type"] = "#Thermal.v1_6_0.Fan";
+        //sensor_json["@odata.type"] = "#Thermal.v1_6_0.Fan";
     	sensor_json["PhysicalContext"] = "Fan";
         setLedState(sensor_json, inventoryItem);
         forceToInt = true;
@@ -948,7 +948,7 @@
     {
         unit = "/Reading"_json_pointer;
         sensor_json["ReadingUnits"] = "Percent";
-        sensor_json["@odata.type"] = "#Thermal.v1_6_0.Fan";
+        //sensor_json["@odata.type"] = "#Thermal.v1_6_0.Fan";
     	sensor_json["PhysicalContext"] = "Fan";
         setLedState(sensor_json, inventoryItem);
         forceToInt = true;
@@ -956,7 +956,7 @@
     else if (sensorType == "voltage")
     {
         unit = "/ReadingVolts"_json_pointer;
-        sensor_json["@odata.type"] = "#Power.v1_6_0.Voltage";
+        //sensor_json["@odata.type"] = "#Power.v1_6_0.Voltage";
     	sensor_json["PhysicalContext"] = "VoltageRegulator";
     }
     else if (sensorType == "power")
@@ -966,7 +966,7 @@
 
         if (!sensorName.compare("total_power"))
         {
-            sensor_json["@odata.type"] = "#Power.v1_6_0.PowerControl";
+            //sensor_json["@odata.type"] = "#Power.v1_6_0.PowerControl";
             // Put multiple "sensors" into a single PowerControl, so have
             // generic names for MemberId and Name. Follows Redfish mockup.
             sensor_json["MemberId"] = "0";
@@ -3178,7 +3178,7 @@
                             return;
                             }
 
-                        aResp->res.jsonValue["@odata.type"] = "#SensorCollection.v1_0_0.ResourceCollection";
+                        aResp->res.jsonValue["@odata.type"] = "#SensorCollection.SensorCollection";
                         aResp->res.jsonValue["@odata.id"] = "/redfish/v1/Chassis/"+chassisId+"/Sensors";
                         aResp->res.jsonValue["@odata.context"] =
                                 "/redfish/v1/$metadata#SensorCollection.SensorCollection";
diff -Naur b/redfish-core/lib/systems.hpp a/redfish-core/lib/systems.hpp
--- b/redfish-core/lib/systems.hpp	2020-12-01 19:02:06.562506040 +0530
+++ a/redfish-core/lib/systems.hpp	2020-12-01 18:52:56.046665716 +0530
@@ -1904,7 +1904,7 @@
     void doGet(crow::Response& res, const crow::Request& req,
                const std::vector<std::string>& params) override
     {
-        res.jsonValue["@odata.type"] = "#ComputerSystem.v1_10_0.ComputerSystem";
+        res.jsonValue["@odata.type"] = "#ComputerSystem.v1_11_0.ComputerSystem";
         res.jsonValue["@odata.context"] = 
 	    "/redfish/v1/$metadata#ComputerSystem.ComputerSystem";
         res.jsonValue["Name"] = "system";
@@ -1924,8 +1924,10 @@
         res.jsonValue["Storage"] = {
             {"@odata.id", "/redfish/v1/Systems/system/Storage"}};
 
-        res.jsonValue["BootOptions"] = {
-            {"@odata.id", "/redfish/v1/Systems/system/BootOptions"}};
+
+	res.jsonValue["Boot"]["BootOptions"] = {{"@odata.id", "/redfish/v1/Systems/system/BootOptions"}};
+	
+        //res.jsonValue["BootOptions"] = {{"@odata.id", "/redfish/v1/Systems/system/BootOptions"}};
 
 
         res.jsonValue["SecureBoot"] = {
@@ -2357,7 +2359,7 @@
 
 	    res.jsonValue["@odata.id"] = "/redfish/v1/Systems/system/BootOptions/"  + BootoptName;
 	    res.jsonValue["Name"] = BootoptName;
-
+	    res.jsonValue["Id"] = BootoptName;
 
 	    std::shared_ptr<AsyncResp> aResp = std::make_shared<AsyncResp>(res);
 
@@ -2432,10 +2434,18 @@
 					    std::string res = "";
 					    if (property.first == "BootOptionEnabled")
 					    {
-					    const std::string *bootoptionenabled = std::get_if<std::string>(&property.second);
-					    aResp->res.jsonValue["BootOptionEnabled"] = *bootoptionenabled;
+					    	const bool *bootoptionenabled = std::get_if<bool>(&property.second);
+					    	aResp->res.jsonValue["BootOptionEnabled"] = *bootoptionenabled;
+
 
 					    }
+					    else if (property.first == "BootOptionReference")
+					    {
+
+						const std::string *bootoptionreference = std::get_if<std::string>(&property.second);
+						aResp->res.jsonValue["BootOptionReference"] = *bootoptionreference;
+
+					    }	    
 
 					    else if (property.first == "DisplayName")
 					    {
@@ -2445,7 +2455,8 @@
 					    else if ((property.first == "UefiDevPath") )
 					    {
 						    const std::string *uefidevpath = std::get_if<std::string>(&property.second);
-						    aResp->res.jsonValue["UefiDevPath"] = *uefidevpath;
+						    aResp->res.jsonValue["UefiDevicePath"] = *uefidevpath;	
+						    		          
 
 					    }
 
@@ -2697,12 +2708,22 @@
 					    else if (property.first == "IPV4Addr")
 					    {
 						    const std::string *ipv4addr = std::get_if<std::string>(&property.second);
-						    aResp->res.jsonValue["IPv4Addresses"] = *ipv4addr;
+						    //aResp->res.jsonValue["IPv4Addresses"] = *ipv4addr;
+						    aResp->res.jsonValue["IPv4Addresses"] = nlohmann::json::array();
+						    auto addrobj = nlohmann::json::object();
+						    addrobj["Address"] = *ipv4addr;
+						    addrobj["SubnetMask"] = "0.0.0.0";
+						    aResp->res.jsonValue["IPv4Addresses"].push_back(addrobj);
 					    }
 					    else if ((property.first == "IPV6Addr") )
 					    {
 						    const std::string *ipv6addr = std::get_if<std::string>(&property.second);
-						    aResp->res.jsonValue["IPv6Addresses"] = *ipv6addr;
+						    //aResp->res.jsonValue["IPv6Addresses"] = *ipv6addr;
+                                                    aResp->res.jsonValue["IPv6Addresses"] = nlohmann::json::array();
+                                                    auto addrobjipv6 = nlohmann::json::object();
+                                                    addrobjipv6["Address"] = *ipv6addr;
+                                                    addrobjipv6["AddressOrigin"] = "Static";
+                                                    aResp->res.jsonValue["IPv6Addresses"].push_back(addrobjipv6);						    
 
 					    }
 
@@ -3106,7 +3127,7 @@
 			const std::string &StorageDriveName = params[1];
 
 
-			res.jsonValue["@odata.type"] = "#Drive.v1_7_0.Drive	";
+			res.jsonValue["@odata.type"] = "#Drive.v1_7_0.Drive";
 			res.jsonValue["@odata.context"] = "/redfish/v1/$metadata#Drive.Drive";
 			res.jsonValue["@odata.id"] =
 				"/redfish/v1/Systems/system/Storage/" + StorageunitName +  "/Drives/" + StorageDriveName ;
