diff -Naur a/redfish-core/lib/log_services.hpp b/redfish-core/lib/log_services.hpp
--- a/redfish-core/lib/log_services.hpp	2020-12-16 17:55:26.944708833 +0530
+++ b/redfish-core/lib/log_services.hpp	2020-12-16 17:58:02.680059550 +0530
@@ -742,6 +742,13 @@
             }
         }
 
+	static const std::string thresholdLogFile = "/var/event/threshold";
+       if(std::filesystem::exists(thresholdLogFile))
+       {
+           std::error_code ec;
+           std::filesystem::remove(thresholdLogFile, ec);
+       }
+
         // Clear the EventLog by deleting the log files
         std::vector<std::filesystem::path> redfishLogFiles;
         if (getRedfishLogFiles(redfishLogFiles))
