diff -Naur a/include/nbd_proxy.hpp b/include/nbd_proxy.hpp
--- a/include/nbd_proxy.hpp	2021-04-27 08:12:31.309044967 +0530
+++ b/include/nbd_proxy.hpp	2021-04-27 08:13:56.744773588 +0530
@@ -35,7 +35,7 @@
 
 using boost::asio::local::stream_protocol;
 
-static constexpr auto nbdBufferSize = 131088;
+static constexpr auto nbdBufferSize = 262176;
 static const char* requiredPrivilegeString = "ConfigureManager";
 
 struct NbdProxyServer : std::enable_shared_from_this<NbdProxyServer>
