diff -Naur a/redfish-core/lib/network_protocol.hpp b/redfish-core/lib/network_protocol.hpp
--- a/redfish-core/lib/network_protocol.hpp	2021-03-01 21:34:15.387758985 +0530
+++ b/redfish-core/lib/network_protocol.hpp	2021-03-01 21:19:30.647835284 +0530
@@ -155,7 +155,7 @@
     {
         std::string hostName;
 
-        std::array<char, HOST_NAME_MAX> hostNameCStr;
+        std::array<char, HOST_NAME_MAX+1> hostNameCStr;
         if (gethostname(hostNameCStr.data(), hostNameCStr.size()) == 0)
         {
             hostName = hostNameCStr.data();
