diff -Naur a/redfish-core/lib/account_service.hpp b/redfish-core/lib/account_service.hpp
--- a/redfish-core/lib/account_service.hpp	2021-03-02 14:33:23.956288223 +0530
+++ b/redfish-core/lib/account_service.hpp	2021-03-02 14:30:47.890561990 +0530
@@ -1639,8 +1639,11 @@
                                     getRoleIdFromPrivilege(*userPrivPtr);
                                 if (role.empty())
                                 {
-                                    BMCWEB_LOG_ERROR << "Invalid user role";
-                                    messages::internalError(asyncResp->res);
+                                    //BMCWEB_LOG_ERROR << "Invalid user role";
+                                    //messages::internalError(asyncResp->res);
+				    //ipmi allows for incomplete users with no user privilage
+                                    BMCWEB_LOG_DEBUG << "No user role";
+
                                     return;
                                 }
                                 asyncResp->res.jsonValue["RoleId"] = role;
