diff -Naur a/redfish-core/lib/account_service.hpp b/redfish-core/lib/account_service.hpp
--- a/redfish-core/lib/account_service.hpp	2021-03-02 15:21:50.053776953 +0530
+++ b/redfish-core/lib/account_service.hpp	2021-03-02 15:10:33.386645760 +0530
@@ -1905,6 +1905,21 @@
             return;
         }
 
+	//added ami osp 2.1 patch 
+        std::vector<const std::string*> sessionIds =
+             crow::persistent_data::SessionStore::getInstance().getUniqueIds(false, crow::persistent_data::PersistenceType::TIMEOUT);
+         for(auto n : sessionIds)
+         {
+             auto session = crow::persistent_data::SessionStore::getInstance().getSessionByUid(*n);
+             if(session->username == params[0])
+             {
+                 messages::internalError(asyncResp->res);
+                 return;
+             }
+        }
+
+
+
         const std::string userPath = "/xyz/openbmc_project/user/" + params[0];
 
         crow::connections::systemBus->async_method_call(
