diff -Naur a/redfish-core/lib/log_services.hpp b/redfish-core/lib/log_services.hpp
--- a/redfish-core/lib/log_services.hpp	2021-03-02 17:06:30.997313897 +0530
+++ b/redfish-core/lib/log_services.hpp	2021-03-02 17:03:08.260026862 +0530
@@ -410,7 +410,7 @@
 static bool
     getRedfishLogFiles(std::vector<std::filesystem::path>& redfishLogFiles)
 {
-    static const std::filesystem::path redfishLogDir = "/var/log";
+    static const std::filesystem::path redfishLogDir = "/var/sellog";
     static const std::string redfishLogFilename = "redfish";
 
     // Loop through the directory looking for redfish log files
