diff -Naur a/http/websocket.h b/http/websocket.h
--- a/http/websocket.h	2021-03-03 13:24:42.728805833 +0530
+++ b/http/websocket.h	2021-03-03 13:36:38.162416265 +0530
@@ -76,6 +76,9 @@
         closeHandler(std::move(close_handler)),
         errorHandler(std::move(error_handler)), session(reqIn.session)
     {
+        ws.set_option(boost::beast::websocket::stream_base::
+                        timeout::suggested(boost::beast::role_type::server));
+    
         BMCWEB_LOG_DEBUG << "Creating new connection " << this;
     }
 
