diff -Naur "a/configurations/Flextronics S-1100ADU00-201 PSU.json" "b/configurations/Flextronics S-1100ADU00-201 PSU.json"
--- "a/configurations/Flextronics S-1100ADU00-201 PSU.json"	2021-02-21 00:26:44.175744345 +0530
+++ "b/configurations/Flextronics S-1100ADU00-201 PSU.json"	2021-02-21 00:27:19.892032282 +0530
@@ -36,28 +36,13 @@
                 ],
 		"SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ]
             },
@@ -89,28 +74,13 @@
                 ],
 		"SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ]
             },
@@ -150,28 +120,13 @@
                 "Name": "PSU$ADDRESS % 4 + 1",
 		"SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
                 "Thresholds": [
diff -Naur "a/configurations/R1000 Chassis.json" "b/configurations/R1000 Chassis.json"
--- "a/configurations/R1000 Chassis.json"	2021-02-21 00:26:44.179744377 +0530
+++ "b/configurations/R1000 Chassis.json"	2021-02-21 00:27:19.892032282 +0530
@@ -8,28 +8,13 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -78,28 +63,13 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -126,28 +96,13 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -174,28 +129,13 @@
             "EvStat": 231,
 	     "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -222,28 +162,13 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -270,28 +195,13 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -320,28 +230,13 @@
             "EvStat": 231,
             "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -368,28 +263,13 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -416,28 +296,13 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -464,28 +329,13 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -512,28 +362,13 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x04",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 52,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 52
                 }
             ],
             "Thresholds": [
@@ -587,28 +422,13 @@
             "QMin": 15,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
diff -Naur "a/configurations/WFT Baseboard.json" "b/configurations/WFT Baseboard.json"
--- "a/configurations/WFT Baseboard.json"	2021-02-21 00:26:44.183744410 +0530
+++ "b/configurations/WFT Baseboard.json"	2021-02-21 00:27:19.892032282 +0530
@@ -9,29 +9,14 @@
             "ScaleFactor": 0.1124,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -69,29 +54,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -129,29 +99,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -190,29 +145,14 @@
             "ScaleFactor": 0.1124,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -251,29 +191,14 @@
             "ScaleFactor": 0.7505,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -313,29 +238,14 @@
             "ScaleFactor": 0.4107,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -380,29 +290,14 @@
             "ScaleFactor": 0.3333,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -443,29 +338,14 @@
             "ScaleFactor": 0.7505,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -506,29 +386,14 @@
             "ScaleFactor": 0.7505,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -568,29 +433,14 @@
             "PowerState": "On",
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -630,29 +480,14 @@
             "PowerState": "On",
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -692,29 +527,14 @@
             "PowerState": "On",
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -754,29 +574,14 @@
             "PowerState": "On",
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -816,29 +621,14 @@
             "PowerState": "On",
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -878,29 +668,14 @@
             "PowerState": "On",
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -938,29 +713,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x02",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 38,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 38
                 }
             ],
             "Thresholds": [
@@ -1153,29 +913,14 @@
             "PowerState": "BiosPost",
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -1215,29 +960,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -1277,29 +1007,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -1339,29 +1054,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -1401,29 +1101,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -1463,29 +1148,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -1525,29 +1195,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -1587,29 +1242,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -1649,29 +1289,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -1711,29 +1336,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -2192,29 +1802,14 @@
             "Name1": "Right Rear TMP421 Internal Temp",
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -2253,29 +1848,14 @@
             "EvStat": 231,
             "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -2314,29 +1894,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -2375,29 +1940,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
@@ -2436,29 +1986,14 @@
             "EvStat": 231,
 	    "SdrInfo": [
                 {
-                        "SdrVersion": "0x51",
                         "RecordType": "0x01",
-                        "RecordLen": 0,
-                        "OwnerId": 0,
-                        "OwnerLun": "0x20",
                         "EntityId": 0,
                         "EntityInstance": 1,
                         "SensorInit": "0xF7",
                         "SensorCap": "0x48",
                         "SensorType" : "0x01",
                         "EventType" : "0x01",
-                        "SensorUnit1" : 0,
-                        "SensorUnit2" : 35,
-                        "ModifierUnit": 0,
-                        "Linearization": 0,
-                        "NominalReading": "0x97",
-                        "NormalMax" : "0xC5",
-                        "NormalMin" : "0x8b",
-                        "MaxReading" : "0xff",
-                        "MinReading" : 0,
-                        "PositiveHysteris" : 0,
-                        "NegativeHysteresis" : 0,
-                        "IdStringLen" : 0
+                        "SensorUnit2" : 35
                 }
             ],
             "Thresholds": [
