diff -Naur a/src/FruDevice.cpp b/src/FruDevice.cpp
--- a/src/FruDevice.cpp	2021-05-13 18:23:13.169142825 +0530
+++ b/src/FruDevice.cpp	2021-05-13 18:23:48.540948540 +0530
@@ -58,7 +58,7 @@
 namespace fs = std::filesystem;
 static constexpr bool DEBUG = false;
 static size_t UNKNOWN_BUS_OBJECT_COUNT = 0;
-constexpr size_t MAX_FRU_SIZE = 512;
+constexpr size_t MAX_FRU_SIZE = 1024;
 constexpr size_t MAX_EEPROM_PAGE_INDEX = 255;
 constexpr size_t busTimeoutSeconds = 5;
 constexpr size_t fruBlockSize = 8; // FRU areas are measured in 8-byte blocks
