From 1156621889e874f12cbe38571d52ae2d4b74b742 Mon Sep 17 00:00:00 2001
From: choppadk <choppadk@ami.com>
Date: Mon, 7 Nov 2022 04:40:02 -0500
Subject: [PATCH] aowanda dts makefile

---
 arch/arm/boot/dts/Makefile | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index b21b3a646..4544a97c9 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -1302,4 +1302,5 @@ dtb-$(CONFIG_ARCH_ASPEED) += \
 	aspeed-bmc-opp-witherspoon.dtb \
 	aspeed-bmc-opp-zaius.dtb \
 	aspeed-bmc-portwell-neptune.dtb \
-	aspeed-bmc-quanta-q71l.dtb
+	aspeed-bmc-quanta-q71l.dtb \
+	aspeed-bmc-mitac-aowanda.dtb
-- 
2.25.1

