--- a/arch/arm/boot/dts/aspeed-g6-ami.dtsi	2020-09-01 16:29:12.463128231 +0800
+++ b/arch/arm/boot/dts/aspeed-g6-ami.dtsi	2020-09-16 15:01:29.529020711 +0800
@@ -717,8 +717,8 @@
 				clocks = <&syscon ASPEED_CLK_GATE_UART4CLK>;
 				resets = <&lpc_reset 7>;
 				no-loopback-test;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_txd4_default &pinctrl_rxd4_default>;
+/*				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_txd4_default &pinctrl_rxd4_default>;*/
 				status = "disabled";
 			};
 
--- a/arch/arm/boot/dts/aspeed-g6-ami.dtsi	2021-05-05 12:33:19.709507240 -0400
+++ b/arch/arm/boot/dts/aspeed-g6-ami.dtsi	2021-05-05 12:43:48.720464581 -0400
@@ -828,8 +828,8 @@
 				clocks = <&syscon ASPEED_CLK_GATE_UART3CLK>;
 				resets = <&lpc_reset 6>;
 				no-loopback-test;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_txd3_default &pinctrl_rxd3_default>;
+				/*pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_txd3_default &pinctrl_rxd3_default>;*/
 				status = "disabled";
 			};
 
