From 207d35c2459d5fb9fddf7495e70d10a564a5a6da Mon Sep 17 00:00:00 2001
From: pranoygoru <gorumanip@amiindia.co.in>
Date: Tue, 2 Mar 2021 17:24:37 +0530
Subject: [PATCH] replace sdbusplus::message::variant

---
 psu-manager/include/utility.hpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/utility.hpp b/include/utility.hpp
index 33bca75..016ad76 100644
--- a/include/utility.hpp
+++ b/include/utility.hpp
@@ -41,7 +41,7 @@ using GetSubTreeType = std::vector<
 
 constexpr std::chrono::microseconds dbusTimeout(5000);
 
-using Value = sdbusplus::message::variant<bool, uint8_t, int16_t, uint16_t,
+using Value = std::variant<bool, uint8_t, int16_t, uint16_t,
                                           int32_t, uint32_t, int64_t, uint64_t,
                                           double, std::string>;
 
