--- a/watchdog.cpp	2022-11-09 04:40:22.022754508 -0500
+++ phosphor-watchdog/watchdog.cpp	2022-11-09 04:44:00.715249958 -0500
@@ -285,7 +285,7 @@
     static constexpr auto SEN_WATCHDOG_SERVICE =
         "xyz.openbmc_project.Eventsensor";
     static constexpr auto SEN_WATCHDOG_ROOT =
-        "/xyz/openbmc_project/sensors/eventsensor/WATCHDOG2";
+        "/xyz/openbmc_project/sensors/watchdog2/IPMI_Watchdog";
     static constexpr auto SEN_WATCHDOG_INTERFACE =
         "xyz.openbmc_project.Sensor.Discrete.State";
     static constexpr auto SEN_INTERFACE = "org.freedesktop.DBus.Properties";
