diff -Naur a/src/sensorcommands.cpp b/src/sensorcommands.cpp
--- a/src/sensorcommands.cpp	2020-08-30 21:39:50.716642667 +0530
+++ b/src/sensorcommands.cpp	2020-08-30 21:41:46.856693149 +0530
@@ -1636,6 +1636,10 @@
         }
 
         name.resize(FULL_RECORD_ID_STR_MAX_LENGTH);
+	if(name.back() == ' '){
+           name.pop_back();
+        }
+
     }
     record.body.id_string_info = name.size();
     std::strncpy(record.body.id_string, name.c_str(),
