diff -Naur a/ipmi-whitelist.conf b/ipmi-whitelist.conf
--- a/ipmi-whitelist.conf	2021-03-15 00:26:48.172074420 +0530
+++ b/ipmi-whitelist.conf	2021-03-15 00:27:34.444110501 +0530
@@ -154,6 +154,12 @@
 0x2c:0x03:0xff7f   //<Group Extension>:<Get Power Limit>
 0x2c:0x04:0x7f7f   //<Group Extension>:<Set Power Limit>
 0x2c:0x05:0x7f7f   //<Group Extension>:<Activate/Deactivate Power Limit>
+0x2c:0x06:0x7f7f   //<Group Extension>:<Get Asset Tag>
+0x2c:0x07:0x7f7f   //<Group Extension>:<Get DCMI Sensor Info>
+0x2c:0x08:0x7f7f   //<Group Extension>:<Set Asset Tag>
+0x2c:0x10:0x7f7f   //<Group Extension>:<Get Temperature Reading>
+0x2c:0x12:0x7f7f   //<Group Extension>:<Set Dcmi Configuration Parameters>
+0x2c:0x13:0x7f7f   //<Group Extension>:<Get Dcmi Configuration Parameters>
 0x2c:0x1f:0xff7f   //<Group Extension>:<Get CPU PECI Package Config Data>
 0x2c:0x20:0xff7f   //<Group Extension>:<Get MDR Data Region Status>
 0x2c:0x21:0x7f7f   //<Group Extension>:<Get MDR Region Update Complete>
