diff -Naur a/include/storagecommands.hpp b/include/storagecommands.hpp
--- a/include/storagecommands.hpp	2020-08-30 22:08:01.684791312 +0530
+++ b/include/storagecommands.hpp	2020-08-30 22:11:53.852811488 +0530
@@ -71,6 +71,7 @@
     amps = 0x5,
     watts = 0x6,
     rpm = 0x12,
+    cfm = 0x11,
 };
 
 enum class IPMINetfnStorageCmds : ipmi_cmd_t
diff -Naur a/src/sensorcommands.cpp b/src/sensorcommands.cpp
--- a/src/sensorcommands.cpp	2020-08-30 22:12:40.591674619 +0530
+++ b/src/sensorcommands.cpp	2020-08-30 22:14:10.004805024 +0530
@@ -73,6 +73,7 @@
 };
 const static boost::container::flat_map<const char*, SensorUnits, CmpStr>
     sensorUnits{{{"temperature", SensorUnits::degreesC},
+	         {"cfm", SensorUnits::cfm},
                  {"voltage", SensorUnits::volts},
                  {"current", SensorUnits::amps},
                  {"fan_tach", SensorUnits::rpm},
