diff -Naur a/include/sensorcommands.hpp b/include/sensorcommands.hpp
--- a/include/sensorcommands.hpp	2020-08-30 22:22:10.383006543 +0530
+++ b/include/sensorcommands.hpp	2020-08-30 22:23:10.109347440 +0530
@@ -144,7 +144,8 @@
 
     if (sensorTree.size() < (sensnum + 1))
     {
-        return IPMI_CC_INVALID_FIELD_REQUEST;
+        //return IPMI_CC_INVALID_FIELD_REQUEST;
+	return IPMI_CC_SENSOR_INVALID;
     }
 
     uint8_t sensorIndex = sensnum;
