--- a/src/sensorcommands.cpp	2021-03-30 20:52:09.710078297 +0530
+++ b/src/sensorcommands.cpp	2021-03-30 20:54:15.783568294 +0530
@@ -425,6 +425,7 @@
 
     double max = 0;
     double min = 0;
+    uint8_t optional = 0;
     getSensorMaxMin(sensorMap, max, min);
 
     int16_t mValue = 0;
@@ -555,8 +556,7 @@
         }
     }
 
-    // no discrete as of today so optional byte is never returned
-    return ipmi::responseSuccess(value, operation, thresholds, std::nullopt);
+    return ipmi::responseSuccess(value, operation, thresholds, optional);
 }
 
 /** @brief implements the Set Sensor threshold command
