diff -Naur a/src/sensorcommands.cpp b/src/sensorcommands.cpp
--- a/src/sensorcommands.cpp	2020-08-30 22:27:48.679560339 +0530
+++ b/src/sensorcommands.cpp	2020-08-30 22:31:30.355105396 +0530
@@ -1772,7 +1772,7 @@
     // <Reserve SDR Repo>
     ipmi::registerHandler(ipmi::prioOemBase, ipmi::netFnSensor,
                           ipmi::sensor_event::cmdReserveDeviceSdrRepository,
-                          ipmi::Privilege::User, ipmiStorageReserveSDR);
+                          ipmi::Privilege::sys_iface, ipmiStorageReserveSDR);
 
     ipmi::registerHandler(ipmi::prioOemBase, ipmi::netFnStorage,
                           ipmi::storage::cmdReserveSdrRepository,
