--- a/src/manufacturingcommands.cpp	2021-03-30 10:01:13.414304177 +0530
+++ b/src/manufacturingcommands.cpp	2021-03-30 10:01:35.370570487 +0530
@@ -788,7 +788,7 @@
                         ipmi::intel::general::cmdSetManufacturingData):
         case makeCmdKey(ipmi::netFnOemOne,
                         ipmi::intel::general::cmdGetManufacturingData):
-        case makeCmdKey(ipmi::netFnStorage, ipmi::storage::cmdWriteFruData):
+      //  case makeCmdKey(ipmi::netFnStorage, ipmi::storage::cmdWriteFruData):
         case makeCmdKey(ipmi::netFnOemTwo, ipmi::intel::platform::cmdClearCMOS):
 
             // Check for Special mode
