--- a/src/sensorcommands.cpp	2021-03-30 21:08:46.588893903 +0530
+++ b/src/sensorcommands.cpp	2021-03-30 21:10:05.901701350 +0530
@@ -970,11 +970,16 @@
         sensorMap.find("xyz.openbmc_project.Sensor.Threshold.Warning");
     auto criticalInterface =
         sensorMap.find("xyz.openbmc_project.Sensor.Threshold.Critical");
+    enabled = static_cast<uint8_t>(
+            IPMISensorEventEnableByte2::sensorScanningEnable);
+    enabled |=
+        static_cast<uint8_t>(IPMISensorReadingByte2::eventMessagesEnable);
+
     if ((warningInterface != sensorMap.end()) ||
         (criticalInterface != sensorMap.end()))
     {
-        enabled = static_cast<uint8_t>(
-            IPMISensorEventEnableByte2::sensorScanningEnable);
+      //  enabled = static_cast<uint8_t>(
+        //    IPMISensorEventEnableByte2::sensorScanningEnable);
         if (warningInterface != sensorMap.end())
         {
             auto& warningMap = warningInterface->second;
