diff -Naur a/src/sensorcommands.cpp b/src/sensorcommands.cpp
--- a/src/sensorcommands.cpp	2020-09-25 14:24:09.180552590 +0530
+++ b/src/sensorcommands.cpp	2020-09-25 16:40:30.513481505 +0530
@@ -1520,24 +1520,6 @@
     record.body.normal_min = NormalMin;
 
     auto typeCstr = type.c_str();
-    //auto findUnits = sensorUnits.find(typeCstr);
-    //if (findUnits != sensorUnits.end())
-   // {
-     //   record.body.sensor_units_2_base =
-      //      static_cast<uint8_t>(findUnits->second);
-   // } // else default 0x0 unspecified
-
-   // record.body.event_reading_type = getSensorEventTypeFromPath(path);
-
-       //auto SensorUnitObj = sensorObject->second.find("Unit");
-    //std::string sensorUnitStr;
-    //if (SensorUnitObj != sensorObject->second.end())
-    //{
-    //   sensorUnitStr = std::get<std::string>(SensorUnitObj->second);
-    //}
-    //if(strcmp(sensorUnitStr.c_str(),"xyz.openbmc_project.Sensor.Value.Unit.Digital") == 0 ||
-    //                 strcmp(sensorUnitStr.c_str(),"xyz.openbmc_project.Sensor.Value.Unit.Discrete") == 0)
-   //{
         //Digital Sensor
         auto SensorDTypeObj = sensorObject->second.find("DType");
         if (SensorDTypeObj != sensorObject->second.end())
@@ -1546,17 +1528,6 @@
           record.body.sensor_units_2_base=0;
 
         }
-        else{
-            record.body.sensor_type=0;
-               auto findUnits = sensorUnits.find(typeCstr);
-        if (findUnits != sensorUnits.end())
-        {
-            record.body.sensor_units_2_base =
-                static_cast<uint8_t>(findUnits->second);
-
-        }
-       }
-        //record.body.sensor_units_2_base=0;
 
         auto SensorEventTypeObj = sensorObject->second.find("EventType");
         if (SensorEventTypeObj != sensorObject->second.end())
@@ -1567,22 +1538,6 @@
         else
             record.body.event_reading_type = getSensorEventTypeFromPath(path);
 
-    //}
-    //else
-    //{
-    //    auto findUnits = sensorUnits.find(typeCstr);
-    //    if (findUnits != sensorUnits.end())
-    //    {
-    //        record.body.sensor_units_2_base =
-    //            static_cast<uint8_t>(findUnits->second);
-
-    //    } // else default 0x0 unspecified
-    //    record.body.event_reading_type = getSensorEventTypeFromPath(path);
-
-    //}
-
-
-
     uint8_t entityId = 0;
     uint8_t entityInstance = 0x01;
 
@@ -1590,8 +1545,6 @@
     // entityInstance
     updateIpmiFromAssociation(path, sensorMap, entityId, entityInstance);
 
-//    record.body.entity_id = entityId;
-  //  record.body.entity_instance = entityInstance;
       record.body.entity_id = EntityId;
       record.body.entity_instance = EntityInstance;
 
