diff -Naur a/src/bridgingcommands.cpp b/src/bridgingcommands.cpp
--- a/src/bridgingcommands.cpp	2020-10-08 18:22:01.690608179 +0530
+++ b/src/bridgingcommands.cpp	2020-10-08 18:07:27.055566486 +0530
@@ -671,9 +671,9 @@
                           ipmi::app::cmdClearMessageFlags,
                           ipmi::Privilege::User, ipmiAppClearMessageFlags);
 
-    ipmi::registerHandler(ipmi::prioOemBase, ipmi::netFnApp,
+/*    ipmi::registerHandler(ipmi::prioOemBase, ipmi::netFnApp,
                           ipmi::app::cmdGetMessageFlags, ipmi::Privilege::User,
-                          ipmiAppGetMessageFlags);
+                          ipmiAppGetMessageFlags); */
 
     ipmi::registerHandler(ipmi::prioOemBase, ipmi::netFnApp,
                           ipmi::app::cmdGetMessage, ipmi::Privilege::User,
diff -Naur a/ipmi-whitelist.conf b/ipmi-whitelist.conf
--- a/ipmi-whitelist.conf	2020-10-08 18:38:48.103331285 +0530
+++ b/ipmi-whitelist.conf	2020-10-08 21:50:00.783798825 +0530
@@ -39,6 +39,7 @@
 0x04:0x30:0x7f7f   //<Sensor/Event>:<Set Sensor Reading and Event Status>
 0x06:0x01:0xffff   //<App>:<Get Device ID>
 0x06:0x02:0x7f7f   //<App>:<Cold Reset>
+0x06:0x03:0x7f7f   //<App>:<Warm Reset>
 0x06:0x04:0xff7f   //<App>:<Get Self Test>
 0x06:0x05:0x7f7f   //<App>:<Manufacturing Test Mode>
 0x06:0x06:0xffff   //<App>:<Set ACPI Power State>
@@ -54,6 +55,7 @@
 0x06:0x33:0xff7f   //<App>:<Get Message>
 0x06:0x34:0x7f7f   //<App>:<Send Message>
 0x06:0x35:0xffff   //<App>:<Read Event Message Buffer>
+0x06:0x36:0xffff   //<App>:<Get BT Interface Capabilities>
 0x06:0x37:0xff7f   //<App>:<Get System GUID>
 0x06:0x38:0xff7f   //<App>:<Get Channel Authentication Capability>
 0x06:0x39:0xff7f   //<App>:<Get Session Challenge>
