diff -Naur a/ipmi-whitelist.conf b/ipmi-whitelist.conf
--- a/ipmi-whitelist.conf	2021-05-13 17:12:09.130676347 +0530
+++ b/ipmi-whitelist.conf	2021-05-13 21:16:43.396595376 +0530
@@ -55,6 +55,7 @@
 0x06:0x33:0x8000   //<App>:<Get Message>
 0x06:0x34:0x7f7f   //<App>:<Send Message>
 0x06:0x35:0x8080   //<App>:<Read Event Message Buffer>
+0x06:0x36:0xff7f   //<App>:<Get BT Interface Capability>
 0x06:0x37:0xff7f   //<App>:<Get System GUID>
 0x06:0x38:0xff7f   //<App>:<Get Channel Authentication Capability>
 0x06:0x39:0xff7f   //<App>:<Get Session Challenge>
