diff -Naur a/src/sensorcommands.cpp b/src/sensorcommands.cpp
--- a/src/sensorcommands.cpp	2020-11-29 16:09:05.095149543 +0530
+++ b/src/sensorcommands.cpp	2020-11-29 16:14:50.228208566 +0530
@@ -898,7 +898,6 @@
     uint8_t sensorTyp;
     uint8_t eventTyp;
     sensorPath = getPathFromSensorNumber(SensorNum);
-    sensorTyp = getSensorTypeFromPath(sensorPath);
 
     std::string connection;
     std::string path;
@@ -914,7 +913,20 @@
     {
         return ipmi::responseResponseError();
     }
+
     auto sensorObject = sensorMap.find("xyz.openbmc_project.Sensor.Value");
+
+    auto SensorTypeObj = sensorObject->second.find("SensorType");
+    if (SensorTypeObj != sensorObject->second.end())
+    {
+        sensorTyp =  std::get<uint8_t>(SensorTypeObj->second);
+
+    }
+    else
+    {
+	sensorTyp = getSensorTypeFromPath(sensorPath);
+    }
+    
     auto SensorEventTypeObj = sensorObject->second.find("EventType");
     if (SensorEventTypeObj != sensorObject->second.end())
     {
