diff -Naur a/src/sensorcommands.cpp b/src/sensorcommands.cpp
--- a/src/sensorcommands.cpp	2020-12-01 16:58:39.137553762 +0530
+++ b/src/sensorcommands.cpp	2020-12-01 17:00:49.583299520 +0530
@@ -1984,7 +1984,7 @@
     // <Get Device SDR Info>
     ipmi::registerHandler(ipmi::prioOpenBmcBase, ipmi::netFnSensor,
                           ipmi::sensor_event::cmdGetDeviceSdrInfo,
-                          ipmi::Privilege::User, ipmiSensorGetDeviceSdrInfo);
+                          ipmi::Privilege::sys_iface, ipmiSensorGetDeviceSdrInfo);
 
     // <Get SDR Repository Info>
     ipmi::registerHandler(ipmi::prioOemBase, ipmi::netFnStorage,
@@ -2010,7 +2010,7 @@
     // <Get Sdr>
     ipmi::registerHandler(ipmi::prioOemBase, ipmi::netFnSensor,
                           ipmi::sensor_event::cmdGetDeviceSdr,
-                          ipmi::Privilege::User, ipmiStorageGetSDR);
+                          ipmi::Privilege::sys_iface, ipmiStorageGetSDR);
 
     ipmi::registerHandler(ipmi::prioOemBase, ipmi::netFnStorage,
                           ipmi::storage::cmdGetSdr, ipmi::Privilege::User,
diff -Naur a/src/storagecommands.cpp b/src/storagecommands.cpp
--- a/src/storagecommands.cpp	2020-12-01 16:58:39.137553762 +0530
+++ b/src/storagecommands.cpp	2020-12-01 17:02:52.369376484 +0530
@@ -539,7 +539,12 @@
         return ipmi::responseInvalidFieldRequest();
     }
 
-    getFru(ctx, fruDeviceId);
+ //   getFru(ctx, fruDeviceId);
+    ipmi::Cc ret = getFru(ctx, fruDeviceId);
+    if (ret != ipmi::ccSuccess)
+    {
+        return ipmi::response(ret);
+    }
 
     constexpr uint8_t accessType =
         static_cast<uint8_t>(GetFRUAreaAccessType::byte);
