diff -Naur a/src/oemcommands.cpp b/src/oemcommands.cpp
--- a/src/oemcommands.cpp	2021-05-18 18:52:43.529976649 +0530
+++ b/src/oemcommands.cpp	2021-05-18 18:54:01.089464548 +0530
@@ -3759,7 +3759,11 @@
     		 {
         		return responseReqDataLenInvalid();
     		 }
-
+		
+		 if (reqData.size() > 64)
+		 {
+			return responseReqDataLenInvalid();
+		 }
 		 std::string sen(reqData.begin(), reqData.end());
 		 sender = sen;
                  mailChk = emailIdCheck(sender);
@@ -3776,6 +3780,11 @@
                  {
                         return responseReqDataLenInvalid();
                  }
+
+		 if (reqData.size() > 64)
+                 {
+                        return responseReqDataLenInvalid();
+                 }
 
 		 std::string reci(reqData.begin(), reqData.end());
 		 rec = reci;
