diff -Naur a/src/sensorcommands.cpp b/src/sensorcommands.cpp
--- a/src/sensorcommands.cpp	2021-05-28 12:29:50.731591088 +0530
+++ b/src/sensorcommands.cpp	2021-05-28 10:51:25.613891066 +0530
@@ -43,7 +43,8 @@
 #include <string>
 #include <utility>
 #include <variant>
-
+#define SDR_DEVICE_SIZE 10
+#define SDR_FULLRECORD_SIZE 16
 namespace ipmi
 {
 using ManagedObjectType =
@@ -1718,6 +1719,8 @@
     uint16_t recordCount =
         sensorTree.size() + fruCount + ipmi::storage::type12Count;
 
+    uint16_t freespace = (SDR_DEVICE_SIZE * 1024) - (recordCount * SDR_FULLRECORD_SIZE);
+
     uint8_t operationSupport = static_cast<uint8_t>(
         SdrRepositoryInfoOps::overflow); // write not supported
 
@@ -1726,7 +1729,7 @@
     operationSupport |= static_cast<uint8_t>(
         SdrRepositoryInfoOps::reserveSDRRepositoryCommandSupported);
     return ipmi::responseSuccess(ipmiSdrVersion, recordCount,
-                                 unspecifiedFreeSpace, sdrLastAdd,
+                                 freespace, sdrLastAdd,
                                  sdrLastRemove, operationSupport);
 }
 
