diff -Naur a/src/sensorcommands.cpp b/src/sensorcommands.cpp
--- a/src/sensorcommands.cpp	2020-12-01 12:58:58.259420749 +0530
+++ b/src/sensorcommands.cpp	2020-12-01 13:06:16.111406056 +0530
@@ -461,14 +461,7 @@
         }
     }
 
-    if (notReading)
-    {
-        operation |= static_cast<uint8_t>(
-            IPMISensorReadingByte2::readingStateUnavailable);
-    }
-
     uint8_t thresholds = 0;
-
    
     //Check Digital Sensor
     /*auto SensorUnitObj = sensorObject->second.find("Unit");
@@ -507,6 +500,11 @@
     }
 
 
+    if (notReading)
+    {
+        operation |= static_cast<uint8_t>(
+            IPMISensorReadingByte2::readingStateUnavailable);
+    }
 
     auto warningObject =
         sensorMap.find("xyz.openbmc_project.Sensor.Threshold.Warning");
