diff -Naur a/src/bridgingcommands.cpp b/src/bridgingcommands.cpp
--- a/src/bridgingcommands.cpp	2020-12-10 15:26:57.224447609 +0530
+++ b/src/bridgingcommands.cpp	2020-12-11 11:12:25.289035956 +0530
@@ -685,7 +685,7 @@
 
     ipmi::registerHandler(ipmi::prioOemBase, ipmi::netFnApp,
                           ipmi::app::cmdReadEventMessageBuffer,
-                          ipmi::Privilege::User, ipmiAppReadEventMessageBuffer);
+                          ipmi::Privilege::sys_iface, ipmiAppReadEventMessageBuffer);
 
     return;
 }
