diff -Naur a/src/storagecommands.cpp b/src/storagecommands.cpp
--- a/src/storagecommands.cpp	2020-12-16 17:48:13.610823187 +0530
+++ b/src/storagecommands.cpp	2020-12-16 17:44:00.429092381 +0530
@@ -1390,6 +1390,13 @@
         }
     }
 
+    static const std::string thresholdLogFile = "/var/event/threshold";
+    if(std::filesystem::exists(thresholdLogFile))
+    {
+	std::error_code ec;
+        std::filesystem::remove(thresholdLogFile, ec);
+    }
+
 
     // Clear the SEL by deleting the log files
     std::vector<std::filesystem::path> redfishLogFiles;
@@ -1402,6 +1409,8 @@
          }
      }
 
+    
+
     // Reload rsyslog so it knows to start new log files
     std::shared_ptr<sdbusplus::asio::connection> dbus = getSdBus();
     sdbusplus::message::message rsyslogReload = dbus->new_method_call(
