diff -Naur a/src/sensorcommands.cpp b/src/sensorcommands.cpp
--- a/src/sensorcommands.cpp	2020-12-18 23:50:45.070035436 +0530
+++ b/src/sensorcommands.cpp	2020-12-19 00:20:38.371086229 +0530
@@ -1195,10 +1195,9 @@
                                std::optional<uint8_t> count)
 {
     uint8_t sdrCount;
-    constexpr uint8_t lunsAndDynamicPopulation = 1;
+    constexpr uint8_t lunsAndDynamicPopulation = 0x81;
     constexpr uint8_t getSdrCount = 0x01;
     constexpr uint8_t getSensorCount = 0x00;
-    sdrLastAdd = 0;
     
     if (sensorTree.empty() && !getSensorSubtree(sensorTree))
 	{
