--- a/ipmi-whitelist.conf	2021-03-03 17:40:16.317238597 +0530
+++ b/ipmi-whitelist.conf	2021-03-03 17:40:41.337394252 +0530
@@ -281,7 +281,7 @@
 0x30:0xb1:0x7f7f   //<Intel General Application>:<Control BMC Services>
 0x30:0xb2:0xff7f   //<Intel General Application>:<Get BMC Service Status>
 0x30:0xb3:0xff7f   //<Intel General Application>:<Get BMC Security Control Mode>
-0x30:0xb4:0x7f7f   //<Intel General Application>:<Set BMC Security Control Mode>
+0x30:0xb4:0xff7f   //<Intel General Application>:<Set BMC Security Control Mode>
 0x30:0xb5:0x7f7f   //<Intel General Application>:<Manufacturing mode Keep Alive>
 0x30:0xbb:0xff7f   //<Intel General Application>:<Get CPLD Revision ID>
 0x30:0xc2:0xff7f   //<Intel General Application>:<Get OEM Extended Sys Info>
