diff -Naur a/image_manager.cpp b/image_manager.cpp
--- a/image_manager.cpp	2020-12-03 23:31:29.473183317 +0530
+++ b/image_manager.cpp	2020-12-03 23:29:13.263902890 +0530
@@ -37,6 +37,7 @@
 using ManifestFail = Software::Image::ManifestFileFailure;
 using UnTarFail = Software::Image::UnTarFailure;
 using InternalFail = Software::Image::InternalFailure;
+using ImageFail = Software::Image::ImageFailure;
 namespace fs = std::filesystem;
 
 struct RemovablePath
@@ -226,6 +227,7 @@
     	tmpDirToRemove.path.clear();
 	log<level::INFO>("Software Object with the same version already exists",
                          entry("VERSION_ID=%s", id.c_str()));
+	report<ImageFailure>(ImageFail::FAIL("Same version image already exist"),ImageFail::PATH(imageDirPath.c_str()));
     }
     return 0;
 }
