diff -Naur a/power-control-x86/src/power_control.cpp b/power-control-x86/src/power_control.cpp
--- a/power-control-x86/src/power_control.cpp	2021-03-27 23:52:09.537776724 +0530
+++ b/power-control-x86/src/power_control.cpp	2021-03-27 23:48:43.792854049 +0530
@@ -73,7 +73,7 @@
 const static constexpr int powerPulseTimeMs = 200;
 const static constexpr int forceOffPulseTimeMs = 15000;
 const static constexpr int resetPulseTimeMs = 500;
-const static constexpr int powerCycleTimeMs = 5000;
+const static constexpr int powerCycleTimeMs = 10000;
 const static constexpr int sioPowerGoodWatchdogTimeMs = 1000;
 const static constexpr int psPowerOKWatchdogTimeMs = 8000;
 const static constexpr int gracefulPowerOffTimeMs = 60000;
