diff --git a/src/ethernet_interface.cpp b/src/ethernet_interface.cpp
index d64944f..2e517b4 100644
--- a/src/ethernet_interface.cpp
+++ b/src/ethernet_interface.cpp
@@ -1062,8 +1062,11 @@ void EthernetInterface::writeConfigurationFile()
             }
         }
     }
-    config.map["IPv6AcceptRA"].emplace_back()["DHCPv6Client"].emplace_back(
-        dhcp6() ? "true" : "false");
+    {
+        auto& ipv6acceptra = config.map["IPv6AcceptRA"].emplace_back();
+        ipv6acceptra["DHCPv6Client"].emplace_back(dhcp6() ? "true" : "false");
+        ipv6acceptra["UseAutonomousPrefix"].emplace_back(dhcp6() ? "true" : "false");        
+    }
     {
         auto& neighbors = config.map["Neighbor"];
         for (const auto& sneighbor : staticNeighbors)
