diff -Naur a/ethernet_interface.cpp b/ethernet_interface.cpp
--- a/ethernet_interface.cpp	2020-11-25 14:33:29.850468303 -0500
+++ b/ethernet_interface.cpp	2020-11-25 14:35:11.606506418 -0500
@@ -940,7 +940,7 @@
         else
         {
 			 if( (EthernetInterfaceIntf::dHCPEnabled() ==
-              EthernetInterface::DHCPConf::v6)  && 
+              EthernetInterface::DHCPConf::none)  && 
               (addr.second->type() == IP::Protocol::IPv4)) 
               {
 					// Process all static addresses
