diff -Naur a/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts b/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts
--- a/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts	2021-02-10 14:13:33.287212583 +0530
+++ b/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts	2021-02-10 14:14:16.099791380 +0530
@@ -65,8 +65,8 @@
         compatible = "iio-hwmon";
         io-channels = <&adc0 0>, <&adc0 1>, <&adc0 2>, <&adc0 3>,
                         <&adc0 4>, <&adc0 5>, <&adc0 6>, <&adc0 7>,
-                        <&adc1 8>, <&adc1 9>, <&adc1 10>, <&adc1 11>,
-                        <&adc1 12>, <&adc1 13>, <&adc1 14>, <&adc1 15>;
+                        <&adc1 0>, <&adc1 1>, <&adc1 2>, <&adc1 3>,
+                        <&adc1 4>, <&adc1 5>, <&adc1 6>, <&adc1 7>;
     };
 
 };
