diff -Naur e/include/dt-bindings/clock/aspeed-clock.h f/include/dt-bindings/clock/aspeed-clock.h
--- e/include/dt-bindings/clock/aspeed-clock.h	2021-10-05 21:28:07.312286214 -0400
+++ f/include/dt-bindings/clock/aspeed-clock.h	2021-10-06 09:52:58.844603637 -0400
@@ -39,6 +39,7 @@
 #define ASPEED_CLK_BCLK			33
 #define ASPEED_CLK_MPLL			34
 #define ASPEED_CLK_24M			35
+#define ASPEED_CLK_MAC1RCLK		36
 
 #define ASPEED_RESET_XDMA		0
 #define ASPEED_RESET_MCTP		1
