--- a/arch/arm/dts/ast2500-evb.dts	2021-02-23 16:23:51.784058115 -0500
+++ b/arch/arm/dts/ast2500-evb.dts	2021-02-23 16:26:01.047862728 -0500
@@ -45,18 +45,21 @@
 };
 
 &mac0 {
-	status = "okay";
-	phy-mode = "rgmii";
+        status = "okay";
 
-	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_mac1link_default &pinctrl_mdio1_default>;
+        phy-mode = "NC-SI";
+        use-ncsi;
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_rmii1_default>;
 };
 
 &mac1 {
-	status = "okay";
-	phy-mode = "rgmii";
-	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_mac2link_default &pinctrl_mdio2_default>;
+        status = "okay";
+
+        phy-mode = "NC-SI";
+        use-ncsi;
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_rmii2_default>;
 };
 
 &fmc {
