--- a/board/aspeed/evb_ast2500/tiogapass.c	2021-03-10 13:43:38.526534040 +0530
+++ b/board/aspeed/evb_ast2500/tiogapass.c	2021-03-10 13:49:02.402153058 +0530
@@ -415,7 +415,7 @@
 	/* Reset WDT2 timer to count down within 60 seconds */                                                                      		\
 //	*((volatile u_long *) (WDT_REG_BASE + WDT2_COUNTER_RELOAD_OFFSET)) = 0x3938700;                                   
 //	*((volatile u_long *) (WDT_REG_BASE + WDT2_COUNTER_RESTART_OFFSET)) = 0x4755;                                   
-                                                                                                         		
+       *((volatile u_long *) SCU_REG_BASE) = 0x1688A8A8;       /* unlock SCU */                                                                                  		
 }
 
 void disable_watchdog2()
