--- a/redfish-core/lib/log_services.hpp	2022-11-09 07:12:10.026981649 -0500
+++ bmcweb/redfish-core/lib/log_services.hpp	2022-11-09 07:16:09.939667176 -0500
@@ -432,7 +432,7 @@
     static constexpr auto SEN_SEL_SERVICE =
         "xyz.openbmc_project.Eventsensor";
     static constexpr auto SEN_SEL_ROOT =
-        "/xyz/openbmc_project/sensors/eventsensor/SEL";
+        "/xyz/openbmc_project/sensors/event_logging_disabled/System_Event_Log";
     static constexpr auto SEN_SEL_INTERFACE =
         "xyz.openbmc_project.Sensor.Discrete.State";
     static constexpr auto SEN_SET_INTERFACE = "org.freedesktop.DBus.Properties";
