--- a/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts	2021-02-23 17:26:18.132502266 -0500
+++ b/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts	2021-02-23 17:23:42.732746411 -0500
@@ -132,6 +132,17 @@
 	use-ncsi;
 };
 
+&mac1 {
+        status = "okay";
+
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_rmii2_default>;
+        clocks = <&syscon ASPEED_CLK_GATE_MAC2CLK>,
+                 <&syscon ASPEED_CLK_MAC2RCLK>;
+        clock-names = "MACCLK", "RCLK";
+        use-ncsi;
+};
+
 &adc {
 	status = "okay";
 };
