diff -Naur a/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts b/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts
--- a/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts	2021-03-12 13:06:45.350107556 +0530
+++ b/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts	2021-03-12 13:06:21.509939480 +0530
@@ -55,6 +55,7 @@
 	status = "okay";
 	flash@0 {
 		status = "okay";
+		label = "bmc";
 		m25p,fast-read;
 #include "openbmc-flash-layout.dtsi"
 	};
