diff -Naur a/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts b/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts
--- a/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts	2021-03-23 15:44:05.700243828 +0530
+++ b/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts	2021-03-23 15:45:10.752758723 +0530
@@ -127,7 +127,7 @@
 	status = "okay";
 	gpio-line-names =
 	/*A0-A7*/	"BMC_CPLD_FPGA_SEL","","","","","","","",
-	/*B0-B7*/	"","BMC_DEBUG_EN","","","","BMC_PPIN","PS_PWROK",
+	/*B0-B7*/	"","BMC_DEBUG_EN","","","","BMC_PPIN","",
 			"IRQ_PVDDQ_GHJ_VRHOT_LVT3",
 	/*C0-C7*/	"","","","","","","","",
 	/*D0-D7*/	"BIOS_MRC_DEBUG_MSG_DIS","BOARD_REV_ID0","",
@@ -181,7 +181,7 @@
 			"","","","",
 	/*AA0-AA7*/	"CPU1_SKTOCC_LVT3","IRQ_SML1_PMBUS_ALERT",
 			"SERVER_POWER_LED","","PECI_MUX_SELECT","UV_HIGH_SET",
-			"","POST_COMPLETE",
+			"","",
 	/*AB0-AB7*/	"IRQ_HSC_FAULT","OCP_MEZZA_PRES","","","","","","",
 	/*AC0-AC7*/	"","","","","","","","";
 };
