--- a/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts	2021-10-25 15:37:02.282222082 -0400
+++ b/arch/arm/boot/dts/aspeed-bmc-facebook-tiogapass.dts	2021-10-25 15:40:11.009747052 -0400
@@ -51,6 +51,10 @@
 
 };
 
+&uart_routing {
+        status = "okay";
+};
+
 &fmc {
 	status = "okay";
 	flash@0 {
@@ -101,7 +105,10 @@
 };
 
 &uart3 {
-	// SoL BMC Console
+	status = "okay";
+};
+
+&uart4 {
 	status = "okay";
 };
 
