--- a/drivers/i2c/busses/i2c-aspeed.c	2020-12-21 18:38:56.238543810 -0500
+++ b/drivers/i2c/busses/i2c-aspeed.c	2020-12-29 18:15:31.217234936 -0500
@@ -133,6 +133,8 @@
 #define AST_I2C_BASE                   0x1E78A000 //in arch/arm/include/asm/soc-ast/hwmap.h
 #endif /* CONFIG_AMII2C */
 
+int flag = 0;
+
 #if IS_ENABLED(CONFIG_AMII2C)
 static u32 i2c_as_io_base_global = AST_I2C_DEV_G_VA_BASE;
 extern struct i2c_as_data as_data_ptr[BUS_COUNT];
@@ -877,11 +879,13 @@
 #endif /* CONFIG_I2C_SLAVE */
 
 	irq_remaining &= ~irq_handled;
-	if (irq_remaining)
+	if (irq_remaining && flag <= 10)
+	{
 		dev_err(bus->dev,
 			"irq handled != irq. expected 0x%08x, but was 0x%08x\n",
 			irq_received, irq_handled);
-
+		flag++;
+	}
 	/* Ack Rx done */
 	if (irq_received & ASPEED_I2CD_INTR_RX_DONE)
 		writel(ASPEED_I2CD_INTR_RX_DONE,
