From 1cf424d2f6cbe2bd8c7ada7fe847df631eb7927e Mon Sep 17 00:00:00 2001
From: sivaprabug <sivaprabug@ami.com>
Date: Fri, 30 Sep 2022 12:28:01 +0530
Subject: [PATCH] invalid password shows wrong error response

Signed-off-by: sivaprabug <sivaprabug@ami.com>
---
 .../UserManagement/ModalUser.vue              | 20 +++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/src/views/SecurityAndAccess/UserManagement/ModalUser.vue b/src/views/SecurityAndAccess/UserManagement/ModalUser.vue
index 90399cf..251b6ae 100644
--- a/src/views/SecurityAndAccess/UserManagement/ModalUser.vue
+++ b/src/views/SecurityAndAccess/UserManagement/ModalUser.vue
@@ -165,6 +165,9 @@
                       })
                     }}
                   </template>
+                  <template v-else-if="!$v.form.password.pattern">
+                    {{ $t('global.form.invalidFormat') }}
+                  </template>
                 </b-form-invalid-feedback>
               </input-password-toggle>
             </b-form-group>
@@ -307,6 +310,9 @@ export default {
           }),
           minLength: minLength(this.passwordRequirements.minLength),
           maxLength: maxLength(this.passwordRequirements.maxLength),
+          pattern: function (pw) {
+            return this.passwordValidation(pw);
+          },
         },
         passwordConfirmation: {
           required: requiredIf(function () {
@@ -384,6 +390,20 @@ export default {
       bvModalEvt.preventDefault();
       this.handleSubmit();
     },
+    passwordValidation(val) {
+      if (
+        /([a-zA-Z0-9])\1\1+|(sshd|bin|daemon|sys|adm|asdfghjk|asdfghjkl|sync|shutdown|halt|mail|news|uucp operator|games|ftp|man|nobody|stunnel4|password|Password|qwertyui|qwertyuiop)+/gi.test(
+          val
+        ) ||
+        /^[!@#$%^&*)(+=._-]+$/gi.test(val) ||
+        this.form.password.indexOf(this.form.username) !== -1 ||
+        /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?!\d)[a-zA-Z\d]$/gi.test(val) ||
+        /^(?:^|\D)(\d{8,20})$/gi.test(val)
+      ) {
+        return false;
+      }
+      return true;
+    },
   },
 };
 </script>
-- 
2.25.1

