From 842d32d16ee31d0733a0212bdfbebd1834ee4dc7 Mon Sep 17 00:00:00 2001
From: Ryon Heichelbech <ryonh@ami.com>
Date: Wed, 13 Oct 2021 16:05:25 -0400
Subject: [PATCH] ast-common: update flash layout for kernel 5.14

---
 include/configs/ast-common.h | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/include/configs/ast-common.h b/include/configs/ast-common.h
index a0243083bd..c5316e08f7 100644
--- a/include/configs/ast-common.h
+++ b/include/configs/ast-common.h
@@ -103,15 +103,15 @@
 #define CONFIG_SYS_MAX_FLASH_BANKS 	(CONFIG_FMC_CS)
 #define CONFIG_SYS_MAX_FLASH_SECT	(8192)		/* max number of sectors on one chip */
 #define CONFIG_ENV_IS_IN_FLASH		1
-#define CONFIG_ENV_ADDR			(AST_FMC_CS0_BASE + 0x60000)
-#define CONFIG_ENV_ADDR_REDUND		(AST_FMC_CS0_BASE + 0x70000)
+#define CONFIG_ENV_ADDR			(AST_FMC_CS0_BASE + 0xe0000)
+#define CONFIG_ENV_ADDR_REDUND		(AST_FMC_CS0_BASE + 0xf0000)
 
-#define CONFIG_ENV_OFFSET		0x60000	/* environment starts here  */
-#define CONFIG_ENV_OFFSET_REDUND	0x70000
+#define CONFIG_ENV_OFFSET		0xe0000	/* environment starts here  */
+#define CONFIG_ENV_OFFSET_REDUND	0xf0000
 #define CONFIG_ENV_SIZE			0x10000	/* Total Size of Environment Sector */
 #define CONFIG_ENV_SIZE_REDUND		0x10000
 
-#define CONFIG_BOOTCOMMAND		"bootm 20080000"
+#define CONFIG_BOOTCOMMAND		"bootm 20100000"
 #define CONFIG_ENV_OVERWRITE
 
 #define ASPEED_ENV_SETTINGS \
-- 
2.17.1

