--- /dev/null
+++ b/redfish-core/lib/sensors.hpp      2022-10-19 11:12:53.075592092 +0530
@@ -64,6 +64,13 @@
     "/xyz/openbmc_project/sensors/count",
     "/xyz/openbmc_project/sensors/airflow",
     "/xyz/openbmc_project/sensors/humidity",
+    "/xyz/openbmc_project/sensors/energy",
+    "/xyz/openbmc_project/sensors/utilization",
+    "/xyz/openbmc_project/sensors/altitude",
+    "/xyz/openbmc_project/sensors/voltage",
+    "/xyz/openbmc_project/sensors/fan_tach",
+    "/xyz/openbmc_project/sensors/temperature",
+    "/xyz/openbmc_project/sensors/fan_pwm",
 #ifdef BMCWEB_NEW_POWERSUBSYSTEM_THERMALSUBSYSTEM
     "/xyz/openbmc_project/sensors/voltage",
     "/xyz/openbmc_project/sensors/fan_tach",
