From 3a0b4a3a7e05a8ee7e6a88795d2fed86911cc116 Mon Sep 17 00:00:00 2001
From: vamsikrishnak <nibinc@ami.com>
Date: Tue, 27 Jun 2023 23:44:35 -0400
Subject: [PATCH 66/66] DateTime patch error

---
 redfish-core/lib/managers.hpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/redfish-core/lib/managers.hpp b/redfish-core/lib/managers.hpp
index af824205..555711ea 100644
--- a/redfish-core/lib/managers.hpp
+++ b/redfish-core/lib/managers.hpp
@@ -2238,7 +2238,8 @@ inline void setDateTime(std::shared_ptr<bmcweb::AsyncResp> aResp,
                 BMCWEB_LOG_DEBUG << "Failed to set elapsed time. "
                                     "DBUS response error "
                                  << ec;
-                messages::internalError(aResp->res);
+                messages::propertyValueResourceConflict(aResp->res,"DateTime",
+				datetime,boost::urls::url_view("/redfish/v1/Managers/bmc/NetworkProtocol"));
                 return;
             }
             aResp->res.jsonValue["DateTime"] = datetime;
-- 
2.17.1

