--- a/include/configs/aspeed-common.h	2021-01-27 11:32:10.983784598 -0500
+++ b/include/configs/aspeed-common.h	2021-01-27 11:33:02.363550275 -0500
@@ -11,12 +11,12 @@
 
 #include <asm/arch/platform.h>
 
-#define CONFIG_BOOTFILE		"all.bin"
+/*#define CONFIG_BOOTFILE		"all.bin"
 
 #define CONFIG_GATEWAYIP	192.168.0.1
 #define CONFIG_NETMASK		255.255.255.0
 #define CONFIG_IPADDR		192.168.0.45
-#define CONFIG_SERVERIP		192.168.0.81
+#define CONFIG_SERVERIP		192.168.0.81 */
 
 #define CONFIG_STANDALONE_LOAD_ADDR 0x83000000
 
