--- kernel-source/drivers/net/ethernet/faraday/ftgmac100.h	2018-12-20 12:15:05.129336254 -0500
+++ kernel-source_mod/drivers/net/ethernet/faraday/ftgmac100.h	2018-12-20 12:14:16.693516129 -0500
@@ -257,7 +257,6 @@
 
 /* Errors we care about for dropping packets */
 #define RXDES0_ANY_ERROR		( \
-	FTGMAC100_RXDES0_RX_ERR		| \
 	FTGMAC100_RXDES0_CRC_ERR	| \
 	FTGMAC100_RXDES0_FTL		| \
 	FTGMAC100_RXDES0_RUNT		| \
