diff -Naur c/drivers/net/ethernet/faraday/ftgmac100.c d/drivers/net/ethernet/faraday/ftgmac100.c
--- c/drivers/net/ethernet/faraday/ftgmac100.c	2021-04-14 12:06:00.885187840 -0400
+++ d/drivers/net/ethernet/faraday/ftgmac100.c	2021-04-14 11:53:00.867839283 -0400
@@ -1842,6 +1842,13 @@
 		priv->rxdes0_edorr_mask = BIT(30);
 		priv->txdes0_edotr_mask = BIT(30);
 		priv->is_aspeed = true;
+		/* Disable ast2600 problematic HW arbitration */
+		if (of_device_is_compatible(np, "aspeed,ast2500-mac") ||
+		    of_device_is_compatible(np, "aspeed,ast2600-mac")) {
+			iowrite32(FTGMAC100_TM_DEFAULT,
+				priv->base +FTGMAC100_OFFSET_TM);
+		}
+
 	} else {
 		priv->rxdes0_edorr_mask = BIT(15);
 		priv->txdes0_edotr_mask = BIT(15);
diff -Naur c/drivers/net/ethernet/faraday/ftgmac100.h d/drivers/net/ethernet/faraday/ftgmac100.h
--- c/drivers/net/ethernet/faraday/ftgmac100.h	2021-04-14 12:06:02.261196840 -0400
+++ d/drivers/net/ethernet/faraday/ftgmac100.h	2021-04-14 11:53:01.267843749 -0400
@@ -170,6 +170,14 @@
 #define FTGMAC100_MACCR_SW_RST		(1 << 31)
 
 /*
+ * test mode control register
+ */
+#define FTGMAC100_TM_RQ_TX_VALID_DIS (1 << 28)
+#define FTGMAC100_TM_RQ_RR_IDLE_PREV (1 << 27)
+#define FTGMAC100_TM_DEFAULT \
+(FTGMAC100_TM_RQ_TX_VALID_DIS | FTGMAC100_TM_RQ_RR_IDLE_PREV)
+
+/*
  * PHY control register
  */
 #define FTGMAC100_PHYCR_MDC_CYCTHR_MASK	0x3f
