--- a/drivers/i2c/busses/Makefile	2021-01-26 17:11:46.266285335 -0500
+++ b/drivers/i2c/busses/Makefile	2021-01-26 17:12:19.622118016 -0500
@@ -35,6 +35,7 @@
 obj-$(CONFIG_I2C_ALTERA)	+= i2c-altera.o
 obj-$(CONFIG_I2C_AMD_MP2)	+= i2c-amd-mp2-pci.o i2c-amd-mp2-plat.o
 obj-$(CONFIG_I2C_ASPEED)        += irq-aspeed-i2c-ic.o i2c-aspeed.o
+obj-$(CONFIG_I2C_NEW_ASPEED)    += i2c-new-aspeed.o ast2600-i2c-global.o
 obj-$(CONFIG_I2C_AT91)		+= i2c-at91.o
 i2c-at91-objs			:= i2c-at91-core.o i2c-at91-master.o
 ifeq ($(CONFIG_I2C_AT91_SLAVE_EXPERIMENTAL),y)
