--- a/drivers/hwmon/aspeed2600-pwm-tacho.c	2021-02-12 13:41:03.937629111 -0500
+++ b/drivers/hwmon/aspeed2600-pwm-tacho.c	2021-02-12 13:41:58.325362714 -0500
@@ -124,7 +124,7 @@
 		.duty_sync_enable = 0,
 		.invert_pin = 0,
 		.rising = 0x00,
-		.falling = 0x0a,
+		.falling = 0xc8,
 		
 	},
 	[1] = {
@@ -135,7 +135,7 @@
 		.duty_sync_enable = 0,
 		.invert_pin = 0,
 		.rising = 0x00,
-		.falling = 0x0a,
+		.falling = 0xc8,
 	},
 	[2] = {
 		.target_freq = 25000,
@@ -145,7 +145,7 @@
 		.duty_sync_enable = 0,
 		.invert_pin = 0,
 		.rising = 0x00,
-		.falling = 0x0a,
+		.falling = 0xc8,
 	},
 	[3] = {
 		.target_freq = 25000,
@@ -155,7 +155,7 @@
 		.duty_sync_enable = 0,
 		.invert_pin = 0,
 		.rising = 0x00,
-		.falling = 0x0a,
+		.falling = 0xc8,
 	},
 	[4] = {
 		.target_freq = 25000,
@@ -165,7 +165,7 @@
 		.duty_sync_enable = 0,
 		.invert_pin = 0,
 		.rising = 0x00,
-		.falling = 0x0a,
+		.falling = 0xc8,
 	},
 	[5] = {
 		.target_freq = 25000,
@@ -175,7 +175,7 @@
 		.duty_sync_enable = 0,
 		.invert_pin = 0,
 		.rising = 0x00,
-		.falling = 0x0a,
+		.falling = 0xc8,
 	},
 	[6] = {
 		.target_freq = 25000,
