diff -Naur a/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts b/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts
--- a/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts	1970-01-01 05:30:00.000000000 +0530
+++ b/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts	2020-06-04 15:19:36.592399576 +0530
@@ -0,0 +1,437 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright 2019 IBM Corp.
+
+/dts-v1/;
+
+#include "aspeed-g6-ami.dtsi"
+#include <dt-bindings/gpio/aspeed-gpio.h>
+#include <dt-bindings/i2c/i2c.h>
+
+/ {
+	model = "AST2600 EVB";
+	compatible = "aspeed,ast2600";
+
+	aliases {
+		serial4 = &uart5;
+	};
+
+	chosen {
+		bootargs = "console=ttyS4,115200n8";
+	};
+
+	memory@80000000 {
+		device_type = "memory";
+		reg = <0x80000000 0x40000000>;
+	};
+
+	reserved-memory {
+                #address-cells = <1>;
+                #size-cells = <1>;
+                ranges;
+
+                gfx_memory: framebuffer {
+                        size = <0x01000000>;
+                        alignment = <0x01000000>;
+                        compatible = "shared-dma-pool";
+                        reusable;
+                };
+
+                video_engine_memory: jpegbuffer {
+                        size = <0x02000000>;    /* 32M */
+                        alignment = <0x01000000>;
+                        compatible = "shared-dma-pool";
+                        reusable;
+                };
+        };
+
+        reserved-memory {
+                #address-cells = <1>;
+                #size-cells = <1>;
+                ranges;
+                video_memory: video {
+                        size = <0x04000000>;
+                        alignment = <0x01000000>;
+                        compatible = "shared-dma-pool";
+                        no-map;
+                };
+        };
+
+        iio-hwmon {
+                compatible = "iio-hwmon";
+                io-channels = <&adc0 0>, <&adc0 1>, <&adc0 2>, <&adc0 3>,
+                        <&adc0 4>, <&adc0 5>, <&adc0 6>, <&adc0 7>,
+                        <&adc1 0>, <&adc1 1>, <&adc1 2>, <&adc1 3>,
+                        <&adc1 4>, <&adc1 5>, <&adc1 6>, <&adc1 7>;
+        };
+
+};
+
+&syscon {
+        uart-clock-high-speed;
+        status = "okay";
+};
+
+&kcs1{
+    status = "okay";
+    kcs_addr = <0xCA0>;
+};
+
+&kcs2{
+    status = "okay";
+    kcs_addr = <0xCA8>;
+};
+
+&kcs3{
+    status = "okay";
+    kcs_addr = <0xCA2>;
+};
+
+&lpc_sio {
+        status = "okay";
+};
+
+&lpc_snoop {
+        snoop-ports = <0x80>;
+        status = "okay";
+};
+
+&mbox {
+        status = "okay";
+};
+
+&mdio0 {
+       status = "okay";
+
+       ethphy0: ethernet-phy@0 {
+               compatible = "ethernet-phy-ieee802.3-c22";
+               reg = <0>;
+       };
+};
+
+&mdio1 {
+	status = "okay";
+
+	ethphy1: ethernet-phy@0 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <0>;
+	};
+};
+
+&mdio2 {
+	status = "okay";
+
+	ethphy2: ethernet-phy@0 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <0>;
+	};
+};
+
+&mdio3 {
+	status = "okay";
+
+	ethphy3: ethernet-phy@0 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <0>;
+	};
+};
+
+&mac0 {
+       status = "okay";
+
+       phy-mode = "rgmii";
+       phy-handle = <&ethphy0>;
+
+       pinctrl-names = "default";
+       pinctrl-0 = <&pinctrl_rgmii1_default>;
+};
+
+&mac1 {
+	status = "okay";
+
+	phy-mode = "rgmii";
+	phy-handle = <&ethphy1>;
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii2_default>;
+};
+
+&mac2 {
+	status = "okay";
+
+	phy-mode = "rgmii";
+	phy-handle = <&ethphy2>;
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii3_default>;
+};
+
+&mac3 {
+	status = "okay";
+
+	phy-mode = "rgmii";
+	phy-handle = <&ethphy3>;
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii4_default>;
+};
+
+&adc0 {
+        status = "okay";
+};
+
+&adc1 {
+        status = "okay";
+};
+
+&uart_routing {
+        status = "okay";
+};
+
+&emmc_controller {
+	status = "okay";
+};
+
+&emmc {
+	non-removable;
+	bus-width = <4>;
+	max-frequency = <52000000>;
+};
+
+&rtc {
+	status = "okay";
+};
+
+&fmc {
+	status = "okay";
+	flash@0 {
+		status = "okay";
+		m25p,fast-read;
+		label = "bmc";
+		spi-max-frequency = <50000000>;
+
+#include "openbmc-flash-layout.dtsi"
+	};
+};
+
+&spi1 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_spi1_default>;
+
+	flash@0 {
+		status = "okay";
+		m25p,fast-read;
+		label = "pnor";
+		spi-max-frequency = <100000000>;
+	};
+};
+
+&uart5 {
+	// Workaround for A0
+	compatible = "snps,dw-apb-uart";
+};
+
+&i2c0 {
+	status = "okay";
+
+	temp@2e {
+		compatible = "adi,adt7490";
+		reg = <0x2e>;
+	};
+};
+
+&i2c1 {
+	status = "okay";
+};
+
+&i2c2 {
+	status = "okay";
+};
+
+&i2c3 {
+	status = "okay";
+};
+
+&i2c4 {
+	status = "okay";
+};
+
+&i2c5 {
+	status = "okay";
+};
+
+&i2c6 {
+	status = "okay";
+};
+
+&i2c7 {
+	status = "okay";
+};
+
+&i2c8 {
+	status = "okay";
+};
+
+&i2c9 {
+	status = "okay";
+};
+
+&i2c12 {
+	status = "okay";
+};
+
+&i2c13 {
+	status = "okay";
+};
+
+&i2c14 {
+	status = "okay";
+};
+
+&i2c15 {
+	status = "okay";
+};
+
+&fsim0 {
+	status = "okay";
+};
+
+&jtag0 {
+	status = "okay";
+};
+
+&jtag1 {
+	status = "okay";
+};
+
+&gfx {
+        status = "okay";
+        memory-region = <&gfx_memory>;
+};
+
+&pwm_tacho {
+        status = "okay";
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_pwm0_default &pinctrl_tach0_default
+                        &pinctrl_pwm1_default &pinctrl_tach1_default
+                        &pinctrl_pwm2_default &pinctrl_tach2_default
+                        &pinctrl_pwm3_default &pinctrl_tach3_default
+                        &pinctrl_pwm4_default &pinctrl_tach4_default
+                        &pinctrl_pwm5_default &pinctrl_tach5_default
+                        &pinctrl_pwm6_default &pinctrl_tach6_default
+                        &pinctrl_pwm7_default &pinctrl_tach7_default
+                        &pinctrl_pwm8g1_default &pinctrl_tach8_default
+                        &pinctrl_pwm9g1_default &pinctrl_tach9_default
+                        &pinctrl_pwm10g1_default &pinctrl_tach10_default
+                        &pinctrl_pwm11g1_default &pinctrl_tach11_default
+                        &pinctrl_pwm12g1_default &pinctrl_tach12_default
+                        &pinctrl_pwm13g1_default &pinctrl_tach13_default
+                        &pinctrl_pwm14g1_default &pinctrl_tach14_default
+                        &pinctrl_pwm15g1_default &pinctrl_tach15_default>;
+
+        fan@0 {
+                reg = <0x00>;
+                period = /bits/ 8 <>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x00>;
+        };
+
+        fan@1 {
+                reg = <0x01>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x01>;
+        };
+
+        fan@2 {
+                reg = <0x02>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x02>;
+        };
+
+        fan@3 {
+                reg = <0x03>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x03>;
+        };
+
+        fan@4 {
+                reg = <0x04>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x04>;
+        };
+
+        fan@5 {
+                reg = <0x05>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x05>;
+        };
+
+        fan@6 {
+                reg = <0x06>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x06>;
+        };
+
+        fan@7 {
+                reg = <0x07>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x07>;
+        };
+
+        fan@8 {
+                reg = <0x08>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x08>;
+        };
+
+        fan@9 {
+                reg = <0x09>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x09>;
+        };
+
+        fan@10 {
+                reg = <0x0a>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x0a>;
+        };
+
+        fan@11 {
+                reg = <0x0b>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x0b>;
+        };
+
+        fan@12 {
+                reg = <0x0c>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x0c>;
+        };
+
+        fan@13 {
+                reg = <0x0d>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x0d>;
+        };
+
+        fan@14 {
+                reg = <0x0e>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x0e>;
+        };
+
+        fan@15 {
+                reg = <0x0f>;
+                cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
+                aspeed,fan-tach-ch = /bits/ 8 <0x0f>;
+        };
+};
+
+&video {
+        status = "okay";
+        memory-region = <&video_engine_memory>;
+};
+
+&vhub {
+        status = "okay";
+};
