--- a/arch/arm/boot/dts/aspeed-ast2500-evb.dts	2019-08-09 11:33:47.900059207 -0400
+++ b/arch/arm/boot/dts/aspeed-ast2500-evb.dts	2019-08-09 11:35:52.810815110 -0400
@@ -96,6 +96,30 @@
 	};
 };
 
+&i2c0{
+	status = "okay";
+};
+
+&i2c1{
+	status = "okay";
+};
+
+&i2c2{
+	status = "okay";
+};
+
+&i2c4{
+	status = "okay";
+};
+
+&i2c5{
+	status = "okay";
+};
+
+&i2c6{
+	status = "okay";
+};
+
 /*
  * Enable port A as device (via the virtual hub) and port B as
  * host by default on the eval board. This can be easily changed
