--- kernel-source/drivers/clk/clk-aspeed.c	2019-11-06 10:13:32.583676496 -0500
+++ kernel-source/drivers/clk/clk-aspeed.c	2019-11-06 10:14:21.391945060 -0500
@@ -703,7 +703,7 @@
 	/* Strap bits 11:9 define the AXI/AHB clock frequency ratio (aka HCLK)*/
 	regmap_read(map, ASPEED_STRAP, &val);
 	val = (val >> 9) & 0x7;
-	WARN(val == 0, "strapping is zero: cannot determine ahb clock");
+	//WARN(val == 0, "strapping is zero: cannot determine ahb clock");
 	div = 2 * (val + 1);
 	hw = clk_hw_register_fixed_factor(NULL, "ahb", "hpll", 0, 1, div);
 	aspeed_clk_data->hws[ASPEED_CLK_AHB] = hw;
