--- a/include/dt-bindings/clock/ast2600-clock.h	2020-01-16 13:37:05.588330257 -0500
+++ b/include/dt-bindings/clock/ast2600-clock.h	2020-01-16 13:40:23.249656476 -0500
@@ -91,6 +91,14 @@
 /* Only list resets here that are not part of a gate */
 #define ASPEED_RESET_ADC		55
 #define ASPEED_RESET_JTAG_MASTER2	54
+#define ASPEED_RESET_I3C7		47
+#define ASPEED_RESET_I3C6               46
+#define ASPEED_RESET_I3C5               45
+#define ASPEED_RESET_I3C4               44
+#define ASPEED_RESET_I3C3               43
+#define ASPEED_RESET_I3C2               42
+#define ASPEED_RESET_I3C1               41
+#define ASPEED_RESET_I3C0               40
 #define ASPEED_RESET_I3C_DMA		39
 #define ASPEED_RESET_PWM		37
 #define ASPEED_RESET_PECI		36
