--- a/arch/arm/boot/dts/aspeed-g6-pinctrl.dtsi	2020-04-14 11:01:49.494742827 +0800
+++ b/arch/arm/boot/dts/aspeed-g6-pinctrl.dtsi	2020-04-13 11:30:06.121860945 +0800
@@ -297,6 +297,16 @@
 		groups = "I2C9";
 	};
 
+	pinctrl_i3c1_default: i3c1_default {
+		function = "I3C1";
+		groups = "I3C1";
+	};
+
+	pinctrl_i3c2_default: i3c2_default {
+		function = "I3C2";
+		groups = "I3C2";
+	};
+
 	pinctrl_i3c3_default: i3c3_default {
 		function = "I3C3";
 		groups = "I3C3";
@@ -1112,6 +1122,11 @@
 		groups = "UART9";
 	};
 
+	pinctrl_usb2ad_default: usb2ad_default {
+		function = "USB2AD";
+		groups = "USBA";
+	};
+
 	pinctrl_vb_default: vb_default {
 		function = "VB";
 		groups = "VB";
