diff -Naur a/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts b/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts
--- a/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts	2021-03-18 17:12:40.688218862 +0530
+++ b/arch/arm/boot/dts/aspeed-ast2600-evb-ami.dts	2021-03-18 17:23:44.545690948 +0530
@@ -329,6 +329,42 @@
         memory-region = <&gfx_memory>;
 };
 
+&gpio0 {
+        status = "okay";
+        /* Enable GPIOP0 and GPIOP2 pass-through by default */
+        /* pinctrl-names = "pass-through";
+        pinctrl-0 = <&pinctrl_thru0_default
+                        &pinctrl_thru1_default>; */
+        gpio-line-names =
+        /*A0-A7*/       "","","","","","","","",
+        /*B0-B7*/       "","","","","","","","",
+        /*C0-C7*/       "","","","","","","","",
+        /*D0-D7*/       "","","","","","","","",
+        /*E0-E7*/       "","","","","","","","",
+        /*F0-F7*/       "","","","","","","","",
+        /*G0-G7*/       "","","","","","","","",
+        /*H0-H7*/       "","","","","","","","",
+        /*I0-I7*/       "","","","","","","","",
+        /*J0-J7*/       "","","","","","","","",
+        /*K0-K7*/       "","","","","","","","",
+        /*L0-L7*/       "","","","","","","","",
+        /*M0-M7*/       "","","","","","","","",
+        /*N0-N7*/       "","","","","","","","",
+        /*O0-O7*/       "","","","","","","","",
+        /*P0-P7*/       "RESET_BUTTON","RESET_OUT","POWER_BUTTON","POWER_OUT","","","","",
+        /*Q0-Q7*/       "","","","","","","","PS_PWROK",
+        /*R0-R7*/       "","","","","","","","",
+        /*S0-S7*/       "","","","","","","","",
+        /*T0-T7*/       "","","","","","","","",
+        /*U0-U7*/       "","","","","","","","",
+        /*V0-V7*/       "","","","","","","","",
+        /*W0-W7*/       "","","","","","","","",
+        /*X0-X7*/       "","","POST_COMPLETE","","","","","",
+        /*Y0-Y7*/       "","","","","","","","",
+        /*Z0-Z7*/       "","","","","","","","";
+};
+
+
 &pwm_tacho {
         status = "okay";
         pinctrl-names = "default";
@@ -339,11 +375,15 @@
                         &pinctrl_pwm4_default &pinctrl_tach4_default
                         &pinctrl_pwm5_default &pinctrl_tach5_default
                         &pinctrl_pwm6_default &pinctrl_tach6_default
-                        &pinctrl_pwm7_default &pinctrl_tach7_default
+                        &pinctrl_tach8_default
+                        &pinctrl_tach9_default
+                        &pinctrl_tach10_default
+                        &pinctrl_tach11_default
+/*                        &pinctrl_pwm7_default &pinctrl_tach7_default
                         &pinctrl_pwm8g1_default &pinctrl_tach8_default
                         &pinctrl_pwm9g1_default &pinctrl_tach9_default
                         &pinctrl_pwm10g1_default &pinctrl_tach10_default
-                        &pinctrl_pwm11g1_default &pinctrl_tach11_default
+                        &pinctrl_pwm11g1_default &pinctrl_tach11_default */
                         &pinctrl_pwm12g1_default &pinctrl_tach12_default
                         &pinctrl_pwm13g1_default &pinctrl_tach13_default
                         &pinctrl_pwm14g1_default &pinctrl_tach14_default
@@ -392,12 +432,12 @@
                 aspeed,fan-tach-ch = /bits/ 8 <0x06>;
         };
 
-        fan@7 {
+/*        fan@7 {
                 reg = <0x07>;
                 cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
                 aspeed,fan-tach-ch = /bits/ 8 <0x07>;
         };
-
+*/
         fan@8 {
                 reg = <0x08>;
                 cooling-levels = /bits/ 8 <125 151 177 203 229 255>;
