--- linux-5.2.8/drivers/spi/Kconfig	2019-08-21 00:37:56.000000000 +0800
+++ new_linux-5.2.8/drivers/spi/Kconfig	2019-10-17 15:07:15.000000000 +0800
@@ -56,6 +56,11 @@
 	  by providing a high-level interface to send memory-like commands.
 
 comment "SPI Master Controller Drivers"
+config SPI_FMC
+	tristate "Aspeed FMC SPI Controller"
+	depends on ARCH_ASPEED
+	help
+	  This selects a driver for the AST FMC SPI Controller
 
 config SPI_ALTERA
 	tristate "Altera SPI Controller"
