From 72b0c9726452953ed653c3a1157e32c72e5c4fe1 Mon Sep 17 00:00:00 2001
From: Tejas Patil <tejaspp@amiindia.co.in>
Date: Tue, 2 Mar 2021 14:56:02 +0000
Subject: [PATCH 60/60] Igore-0x3FF-in-aspeed_adc-driver

---
 drivers/iio/adc/aspeed_adc.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/iio/adc/aspeed_adc.c b/drivers/iio/adc/aspeed_adc.c
index fb30d5be4..2de147c18 100644
--- a/drivers/iio/adc/aspeed_adc.c
+++ b/drivers/iio/adc/aspeed_adc.c
@@ -46,6 +46,9 @@
 #define ASPEED_ADC_INIT_POLLING_TIME	500
 #define ASPEED_ADC_INIT_TIMEOUT		500000
 
+#define ASPEED_ADC_CHANNELS_MAX        16
+#define ASPEED_ADC_RAW_VALUE_MAX       0x3ff
+
 struct aspeed_adc_model_data {
 	const char *model_name;
 	unsigned int min_sampling_rate;	// Hz
@@ -64,6 +67,7 @@ struct aspeed_adc_data {
 	struct clk_hw		*clk_scaler;
 	struct reset_control	*rst;
 	int 				cv;
+	int channel_raw_value[ASPEED_ADC_CHANNELS_MAX];
 };
 
 #define ASPEED_CHAN(_idx, _data_reg_addr) {			\
@@ -116,6 +120,13 @@ static int aspeed_adc_read_raw(struct iio_dev *indio_dev,
 	switch (mask) {
 	case IIO_CHAN_INFO_RAW:
 		*val = readw(data->base + chan->address);
+                if (*val == ASPEED_ADC_RAW_VALUE_MAX) {
+                        *val = data->channel_raw_value[chan->channel];
+                        pr_err("aspeed_adc: channel %d drop invalid raw reading 0x3FF %d\n",
+                                chan->channel, ASPEED_ADC_RAW_VALUE_MAX);
+                } else {
+                        data->channel_raw_value[chan->channel] = *val;
+                }		
 		return IIO_VAL_INT;
 
 	case IIO_CHAN_INFO_SCALE:
@@ -198,6 +209,7 @@ static int aspeed_adc_probe(struct platform_device *pdev)
 	int ret;
 	u32 eng_ctrl = 0;
 	u32 adc_engine_control_reg_val;
+	int i;
 
 	indio_dev = devm_iio_device_alloc(&pdev->dev, sizeof(*data));
 	if (!indio_dev)
@@ -289,6 +301,9 @@ static int aspeed_adc_probe(struct platform_device *pdev)
 
 	data->cv = 0x200 - (readl(data->base + 0x10) & GENMASK(9, 0));
 
+        for (i = 0; i < ASPEED_ADC_CHANNELS_MAX; i++)
+                data->channel_raw_value[i] = ASPEED_ADC_RAW_VALUE_MAX;
+ 
 	writel(eng_ctrl | ASPEED_OPERATION_MODE_NORMAL | 
 			ASPEED_ENGINE_ENABLE | ASPEED_AUTOPENSATING, data->base + ASPEED_REG_ENGINE_CONTROL);
 	printk(KERN_INFO "aspeed_adc: cv %d \n", data->cv);	
-- 
2.17.1

