--- a/drivers/gpio/gpio-aspeed.c	2020-04-26 11:34:44.602376995 -0400
+++ b/drivers/gpio/gpio-aspeed.c	2020-05-01 17:41:16.299040116 -0400
@@ -1141,7 +1141,7 @@
 	const struct of_device_id *gpio_id;
 	struct aspeed_gpio *gpio;
 	int rc, i, banks, err;
-	u32 ngpio;
+	u32 ngpio, base;
 
 	gpio = devm_kzalloc(&pdev->dev, sizeof(*gpio), GFP_KERNEL);
 	if (!gpio)
@@ -1167,8 +1167,12 @@
 	gpio->config = gpio_id->data;
 
 	gpio->chip.parent = &pdev->dev;
+
+	err = of_property_read_u32(pdev->dev.of_node, "base", &base);
+
 	err = of_property_read_u32(pdev->dev.of_node, "ngpios", &ngpio);
 	gpio->chip.ngpio = (u16) ngpio;
+
 	if (err)
 		gpio->chip.ngpio = gpio->config->nr_gpios;
 	gpio->chip.direction_input = aspeed_gpio_dir_in;
@@ -1180,7 +1184,7 @@
 	gpio->chip.set = aspeed_gpio_set;
 	gpio->chip.set_config = aspeed_gpio_set_config;
 	gpio->chip.label = dev_name(&pdev->dev);
-	gpio->chip.base = -1;
+	gpio->chip.base = (u16) base;
 
 	/* Allocate a cache of the output registers */
 	banks = DIV_ROUND_UP(gpio->chip.ngpio, 32);
