--- a/drivers/gpio/gpio-aspeed.c	2020-07-29 12:00:23.708166349 -0400
+++ b/drivers/gpio/gpio-aspeed.c	2020-07-29 12:04:48.623563627 -0400
@@ -1115,8 +1115,8 @@
 
 static const struct aspeed_bank_props ast2600_bank_props[] = {
 	/*     input	  output   */
-	{5, 0xffffffff,  0x0000ffff}, /* U/V/W/X */
-	{6, 0xffff0000,  0x0fff0000}, /* Y/Z */
+	{5, 0xffffffff,  0xffffff00}, /* U/V/W/X */
+	{6, 0x0000ffff,  0x0000ffff}, /* Y/Z */
 	{ },
 };
 
