diff -Naur a/drivers/usb/gadget/function/f_hid.c b/drivers/usb/gadget/function/f_hid.c
--- a/drivers/usb/gadget/function/f_hid.c	2021-03-30 00:22:08.781304719 +0530
+++ b/drivers/usb/gadget/function/f_hid.c	2021-03-30 00:25:15.320288447 +0530
@@ -319,15 +319,17 @@
 {
 	struct f_hidg *hidg = (struct f_hidg *)ep->driver_data;
 	unsigned long flags;
+
 /*
  * Endpoint Request error: -108 is about host os shutdown the USB connection between BMC and HOST.
  * because of BMC is USB device mode, in BMC's point of view, this error is just an information
+ * so commenting the below code
  * */
 
-	if (req->status != 0) {
-		ERROR(hidg->func.config->cdev,
-			"End Point Request failed since HOST OS shutdown USB connection: %d\n", req->status);
-	}
+//	if (req->status != 0) {
+//		ERROR(hidg->func.config->cdev,
+//			"End Point Request failed since HOST OS shutdown USB connection: %d\n", req->status);
+//	}
 
 	spin_lock_irqsave(&hidg->write_spinlock, flags);
 	hidg->write_pending = 0;
