--- a/phosphor-ldap-config/ldap_config.cpp	2020-10-28 12:15:37.959360747 -0400
+++ b/phosphor-ldap-config/ldap_config.cpp	2020-12-02 16:56:40.688582760 -0500
@@ -227,7 +227,7 @@
     if (secureLDAP == true)
     {
         confData << "ssl on\n";
-        confData << "tls_reqcert hard\n";
+        confData << "tls_reqcert allow\n";
         if (fs::is_directory(tlsCacertFile.c_str()))
         {
             confData << "tls_cacertdir " << tlsCacertFile.c_str() << "\n";
