diff -Naur a/group.cpp b/group.cpp
--- a/group.cpp	2021-05-24 11:55:08.976744961 -0400
+++ b/group.cpp	2021-05-24 15:38:24.705523930 -0400
@@ -33,7 +33,7 @@
     // Set OperationalStatus according to the status of asserted property.
     // If the group is asserted, then the functional status is false and
     // vice-versa.
-    manager.setOperationalStatus(path, !value);
+    // manager.setOperationalStatus(path, !value);
 
     // If something does not go right here, then there should be an sdbusplus
     // exception thrown.
